From 40af2074a546a4160bbe40f9c980f1d8399ad182 Mon Sep 17 00:00:00 2001
From: Steve Hay <me@stevenhay.com>
Date: Thu, 9 Mar 2023 13:02:06 +0100
Subject: [PATCH] arm64: dts: meson-g12b-odroid: Add initial support for
 Hardkernel ODROID-N2L

This patch has been modified to revert the dts refactor and apply a
standalone patch fo the n2l

Signed-off-by: Steve Hay <me@stevenhay.com>

ODROID-N2L is a variant SBC in small form factor and some peripherals
are removed from ODROID-N2PLUS based on S922X SoC.

- On-board ethernet is removed
- On-board RTC is removed
- USB 3.0 hub is removed, so one USB 2.0 and one USB 3.0 host ports
  are available
- Huge heatsink is replaced with 40x40mm heatsink, 5V active heatsink
  is recommended or a tall passive sink is optional
- 3.5mm earjack is removed
- IR remote receiver is removed
- MIPI DSI port is added new but not yet supported

Signed-off-by: Dongjin Kim <tobetter@gmail.com>
Link: https://lore.kernel.org/r/20230122-topic-odroid-n2l-upstream-initial-v2-3-8d3fea6d403d@linaro.org
Signed-off-by: Neil Armstrong <neil.armstrong@linaro.org>
---
 .../devicetree/bindings/arm/amlogic.yaml      |    1 +
 arch/arm64/boot/dts/amlogic/Makefile          |    1 +
 .../dts/amlogic/meson-g12b-odroid-n2l.dts     | 3733 +++++++++++++++++
 3 files changed, 3735 insertions(+)
 create mode 100644 arch/arm64/boot/dts/amlogic/meson-g12b-odroid-n2l.dts

diff --git a/Documentation/devicetree/bindings/arm/amlogic.yaml b/Documentation/devicetree/bindings/arm/amlogic.yaml
index 4f6b8a44f28c..b6c3959aacdc 100644
--- a/Documentation/devicetree/bindings/arm/amlogic.yaml
+++ b/Documentation/devicetree/bindings/arm/amlogic.yaml
@@ -181,6 +181,7 @@ properties:
               - dream,dreambox-one
               - dream,dreambox-two
               - hardkernel,odroid-n2
+              - hardkernel,odroid-n2l
               - hardkernel,odroid-n2-plus
               - khadas,vim3
               - ugoos,am6
diff --git a/arch/arm64/boot/dts/amlogic/Makefile b/arch/arm64/boot/dts/amlogic/Makefile
index 83c75550fd88..a4551ddfff46 100644
--- a/arch/arm64/boot/dts/amlogic/Makefile
+++ b/arch/arm64/boot/dts/amlogic/Makefile
@@ -17,6 +17,7 @@ dtb-$(CONFIG_ARCH_MESON) += meson-g12b-gtking-pro.dtb
 dtb-$(CONFIG_ARCH_MESON) += meson-g12b-gtking.dtb
 dtb-$(CONFIG_ARCH_MESON) += meson-g12b-odroid-n2-plus.dtb
 dtb-$(CONFIG_ARCH_MESON) += meson-g12b-odroid-n2.dtb
+dtb-$(CONFIG_ARCH_MESON) += meson-g12b-odroid-n2l.dtb
 dtb-$(CONFIG_ARCH_MESON) += meson-g12b-radxa-zero2.dtb
 dtb-$(CONFIG_ARCH_MESON) += meson-g12b-s922x-khadas-vim3.dtb
 dtb-$(CONFIG_ARCH_MESON) += meson-g12b-ugoos-am6.dtb
diff --git a/arch/arm64/boot/dts/amlogic/meson-g12b-odroid-n2l.dts b/arch/arm64/boot/dts/amlogic/meson-g12b-odroid-n2l.dts
new file mode 100644
index 000000000000..7163c329fcfa
--- /dev/null
+++ b/arch/arm64/boot/dts/amlogic/meson-g12b-odroid-n2l.dts
@@ -0,0 +1,3733 @@
+# 1 "arch/arm64/boot/dts/amlogic/meson-g12b-odroid-n2l.dts"
+# 1 "<built-in>"
+# 1 "<command-line>"
+# 1 "arch/arm64/boot/dts/amlogic/meson-g12b-odroid-n2l.dts"
+
+
+
+
+
+/dts-v1/;
+
+
+# 1 "arch/arm64/boot/dts/amlogic/meson-g12b-a311d.dtsi" 1
+
+
+
+
+
+
+# 1 "arch/arm64/boot/dts/amlogic/meson-g12b.dtsi" 1
+
+
+
+
+
+
+# 1 "arch/arm64/boot/dts/amlogic/meson-g12.dtsi" 1
+
+
+
+
+
+
+# 1 "arch/arm64/boot/dts/amlogic/meson-g12-common.dtsi" 1
+
+
+
+
+
+# 1 "include/dt-bindings/phy/phy.h" 1
+# 7 "arch/arm64/boot/dts/amlogic/meson-g12-common.dtsi" 2
+# 1 "include/dt-bindings/gpio/gpio.h" 1
+# 8 "arch/arm64/boot/dts/amlogic/meson-g12-common.dtsi" 2
+# 1 "include/dt-bindings/clock/g12a-clkc.h" 1
+# 9 "arch/arm64/boot/dts/amlogic/meson-g12-common.dtsi" 2
+# 1 "include/dt-bindings/clock/g12a-aoclkc.h" 1
+# 10 "arch/arm64/boot/dts/amlogic/meson-g12-common.dtsi" 2
+# 1 "include/dt-bindings/interrupt-controller/irq.h" 1
+# 11 "arch/arm64/boot/dts/amlogic/meson-g12-common.dtsi" 2
+# 1 "include/dt-bindings/interrupt-controller/arm-gic.h" 1
+# 12 "arch/arm64/boot/dts/amlogic/meson-g12-common.dtsi" 2
+# 1 "include/dt-bindings/reset/amlogic,meson-g12a-reset.h" 1
+# 13 "arch/arm64/boot/dts/amlogic/meson-g12-common.dtsi" 2
+# 1 "include/dt-bindings/thermal/thermal.h" 1
+# 14 "arch/arm64/boot/dts/amlogic/meson-g12-common.dtsi" 2
+
+/ {
+ interrupt-parent = <&gic>;
+ #address-cells = <2>;
+ #size-cells = <2>;
+
+ aliases {
+  mmc0 = &sd_emmc_b;
+  mmc1 = &sd_emmc_c;
+  mmc2 = &sd_emmc_a;
+ };
+
+ chosen {
+  #address-cells = <2>;
+  #size-cells = <2>;
+  ranges;
+
+  simplefb_cvbs: framebuffer-cvbs {
+   compatible = "amlogic,simple-framebuffer",
+         "simple-framebuffer";
+   amlogic,pipeline = "vpu-cvbs";
+   clocks = <&clkc 168>,
+     <&clkc 53>,
+     <&clkc 58>;
+   status = "disabled";
+  };
+
+  simplefb_hdmi: framebuffer-hdmi {
+   compatible = "amlogic,simple-framebuffer",
+        "simple-framebuffer";
+   amlogic,pipeline = "vpu-hdmi";
+   clocks = <&clkc 168>,
+     <&clkc 53>,
+     <&clkc 58>;
+   status = "disabled";
+  };
+ };
+
+ efuse: efuse {
+  compatible = "amlogic,meson-gxbb-efuse";
+  clocks = <&clkc 106>;
+  #address-cells = <1>;
+  #size-cells = <1>;
+  read-only;
+  secure-monitor = <&sm>;
+ };
+
+ gpu_opp_table: opp-table-gpu {
+  compatible = "operating-points-v2";
+
+  opp-124999998 {
+   opp-hz = /bits/ 64 <124999998>;
+   opp-microvolt = <800000>;
+  };
+  opp-249999996 {
+   opp-hz = /bits/ 64 <249999996>;
+   opp-microvolt = <800000>;
+  };
+  opp-285714281 {
+   opp-hz = /bits/ 64 <285714281>;
+   opp-microvolt = <800000>;
+  };
+  opp-399999994 {
+   opp-hz = /bits/ 64 <399999994>;
+   opp-microvolt = <800000>;
+  };
+  opp-499999992 {
+   opp-hz = /bits/ 64 <499999992>;
+   opp-microvolt = <800000>;
+  };
+  opp-666666656 {
+   opp-hz = /bits/ 64 <666666656>;
+   opp-microvolt = <800000>;
+  };
+  opp-799999987 {
+   opp-hz = /bits/ 64 <799999987>;
+   opp-microvolt = <800000>;
+  };
+ };
+
+ psci {
+  compatible = "arm,psci-1.0";
+  method = "smc";
+ };
+
+ reserved-memory {
+  #address-cells = <2>;
+  #size-cells = <2>;
+  ranges;
+
+
+  secmon_reserved: secmon@5000000 {
+   reg = <0x0 0x05000000 0x0 0x300000>;
+   no-map;
+  };
+
+
+  secmon_reserved_bl32: secmon@5300000 {
+   reg = <0x0 0x05300000 0x0 0x2000000>;
+   no-map;
+  };
+
+  linux,cma {
+   compatible = "shared-dma-pool";
+   reusable;
+   size = <0x0 0x38000000>;
+   alignment = <0x0 0x400000>;
+   linux,cma-default;
+  };
+ };
+
+ sm: secure-monitor {
+  compatible = "amlogic,meson-gxbb-sm";
+ };
+
+ soc {
+  compatible = "simple-bus";
+  #address-cells = <2>;
+  #size-cells = <2>;
+  ranges;
+
+  pcie: pcie@fc000000 {
+   compatible = "amlogic,g12a-pcie", "snps,dw-pcie";
+   reg = <0x0 0xfc000000 0x0 0x400000>,
+         <0x0 0xff648000 0x0 0x2000>,
+         <0x0 0xfc400000 0x0 0x200000>;
+   reg-names = "elbi", "cfg", "config";
+   interrupts = <0 221 4>;
+   #interrupt-cells = <1>;
+   interrupt-map-mask = <0 0 0 0>;
+   interrupt-map = <0 0 0 0 &gic 0 223 4>;
+   bus-range = <0x0 0xff>;
+   #address-cells = <3>;
+   #size-cells = <2>;
+   device_type = "pci";
+   ranges = <0x81000000 0 0 0x0 0xfc600000 0 0x00100000>,
+     <0x82000000 0 0xfc700000 0x0 0xfc700000 0 0x1900000>;
+
+   clocks = <&clkc 48
+      &clkc 45
+      &clkc 201>;
+   clock-names = "general",
+          "pclk",
+          "port";
+   resets = <&reset 12>,
+     <&reset 15>;
+   reset-names = "port",
+          "apb";
+   num-lanes = <1>;
+   phys = <&usb3_pcie_phy 2>;
+   phy-names = "pcie";
+   status = "disabled";
+  };
+
+  ethmac: ethernet@ff3f0000 {
+   compatible = "amlogic,meson-g12a-dwmac",
+         "snps,dwmac-3.70a",
+         "snps,dwmac";
+   reg = <0x0 0xff3f0000 0x0 0x10000>,
+         <0x0 0xff634540 0x0 0x8>;
+   interrupts = <0 8 4>;
+   interrupt-names = "macirq";
+   clocks = <&clkc 38>,
+     <&clkc 2>,
+     <&clkc 13>,
+     <&clkc 2>;
+   clock-names = "stmmaceth", "clkin0", "clkin1",
+          "timing-adjustment";
+   rx-fifo-depth = <4096>;
+   tx-fifo-depth = <2048>;
+   status = "disabled";
+
+   mdio0: mdio {
+    #address-cells = <1>;
+    #size-cells = <0>;
+    compatible = "snps,dwmac-mdio";
+   };
+  };
+
+  apb: bus@ff600000 {
+   compatible = "simple-bus";
+   reg = <0x0 0xff600000 0x0 0x200000>;
+   #address-cells = <2>;
+   #size-cells = <2>;
+   ranges = <0x0 0x0 0x0 0xff600000 0x0 0x200000>;
+
+   hdmi_tx: hdmi-tx@0 {
+    compatible = "amlogic,meson-g12a-dw-hdmi";
+    reg = <0x0 0x0 0x0 0x10000>;
+    interrupts = <0 57 1>;
+    resets = <&reset 19>,
+      <&reset 66>,
+      <&reset 79>;
+    reset-names = "hdmitx_apb", "hdmitx", "hdmitx_phy";
+    clocks = <&clkc 168>,
+      <&clkc 53>,
+      <&clkc 58>;
+    clock-names = "isfr", "iahb", "venci";
+    #address-cells = <1>;
+    #size-cells = <0>;
+    #sound-dai-cells = <0>;
+    status = "disabled";
+
+
+    hdmi_tx_venc_port: port@0 {
+     reg = <0>;
+
+     hdmi_tx_in: endpoint {
+      remote-endpoint = <&hdmi_tx_out>;
+     };
+    };
+
+
+    hdmi_tx_tmds_port: port@1 {
+     reg = <1>;
+    };
+   };
+
+   apb_efuse: bus@30000 {
+    compatible = "simple-bus";
+    reg = <0x0 0x30000 0x0 0x2000>;
+    #address-cells = <2>;
+    #size-cells = <2>;
+    ranges = <0x0 0x0 0x0 0x30000 0x0 0x2000>;
+
+    hwrng: rng@218 {
+     compatible = "amlogic,meson-rng";
+     reg = <0x0 0x218 0x0 0x4>;
+     clocks = <&clkc 27>;
+     clock-names = "core";
+    };
+   };
+
+   acodec: audio-controller@32000 {
+    compatible = "amlogic,t9015";
+    reg = <0x0 0x32000 0x0 0x14>;
+    #sound-dai-cells = <0>;
+    sound-name-prefix = "ACODEC";
+    clocks = <&clkc 36>;
+    clock-names = "pclk";
+    resets = <&reset 61>;
+    status = "disabled";
+   };
+
+   periphs: bus@34400 {
+    compatible = "simple-bus";
+    reg = <0x0 0x34400 0x0 0x400>;
+    #address-cells = <2>;
+    #size-cells = <2>;
+    ranges = <0x0 0x0 0x0 0x34400 0x0 0x400>;
+
+    periphs_pinctrl: pinctrl@40 {
+     compatible = "amlogic,meson-g12a-periphs-pinctrl";
+     #address-cells = <2>;
+     #size-cells = <2>;
+     ranges;
+
+     gpio: bank@40 {
+      reg = <0x0 0x40 0x0 0x4c>,
+            <0x0 0xe8 0x0 0x18>,
+            <0x0 0x120 0x0 0x18>,
+            <0x0 0x2c0 0x0 0x40>,
+            <0x0 0x340 0x0 0x1c>;
+      reg-names = "gpio",
+           "pull",
+           "pull-enable",
+           "mux",
+           "ds";
+      gpio-controller;
+      #gpio-cells = <2>;
+      gpio-ranges = <&periphs_pinctrl 0 0 86>;
+     };
+
+     cec_ao_a_h_pins: cec_ao_a_h {
+      mux {
+       groups = "cec_ao_a_h";
+       function = "cec_ao_a_h";
+       bias-disable;
+      };
+     };
+
+     cec_ao_b_h_pins: cec_ao_b_h {
+      mux {
+       groups = "cec_ao_b_h";
+       function = "cec_ao_b_h";
+       bias-disable;
+      };
+     };
+
+     emmc_ctrl_pins: emmc-ctrl {
+      mux-0 {
+       groups = "emmc_cmd";
+       function = "emmc";
+       bias-pull-up;
+       drive-strength-microamp = <4000>;
+      };
+
+      mux-1 {
+       groups = "emmc_clk";
+       function = "emmc";
+       bias-disable;
+       drive-strength-microamp = <4000>;
+      };
+     };
+
+     emmc_data_4b_pins: emmc-data-4b {
+      mux-0 {
+       groups = "emmc_nand_d0",
+         "emmc_nand_d1",
+         "emmc_nand_d2",
+         "emmc_nand_d3";
+       function = "emmc";
+       bias-pull-up;
+       drive-strength-microamp = <4000>;
+      };
+     };
+
+     emmc_data_8b_pins: emmc-data-8b {
+      mux-0 {
+       groups = "emmc_nand_d0",
+         "emmc_nand_d1",
+         "emmc_nand_d2",
+         "emmc_nand_d3",
+         "emmc_nand_d4",
+         "emmc_nand_d5",
+         "emmc_nand_d6",
+         "emmc_nand_d7";
+       function = "emmc";
+       bias-pull-up;
+       drive-strength-microamp = <4000>;
+      };
+     };
+
+     emmc_ds_pins: emmc-ds {
+      mux {
+       groups = "emmc_nand_ds";
+       function = "emmc";
+       bias-pull-down;
+       drive-strength-microamp = <4000>;
+      };
+     };
+
+     emmc_clk_gate_pins: emmc_clk_gate {
+      mux {
+       groups = "BOOT_8";
+       function = "gpio_periphs";
+       bias-pull-down;
+       drive-strength-microamp = <4000>;
+      };
+     };
+
+     hdmitx_ddc_pins: hdmitx_ddc {
+      mux {
+       groups = "hdmitx_sda",
+         "hdmitx_sck";
+       function = "hdmitx";
+       bias-disable;
+       drive-strength-microamp = <4000>;
+      };
+     };
+
+     hdmitx_hpd_pins: hdmitx_hpd {
+      mux {
+       groups = "hdmitx_hpd_in";
+       function = "hdmitx";
+       bias-disable;
+      };
+     };
+
+
+     i2c0_sda_c_pins: i2c0-sda-c {
+      mux {
+       groups = "i2c0_sda_c";
+       function = "i2c0";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+
+      };
+     };
+
+     i2c0_sck_c_pins: i2c0-sck-c {
+      mux {
+       groups = "i2c0_sck_c";
+       function = "i2c0";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     i2c0_sda_z0_pins: i2c0-sda-z0 {
+      mux {
+       groups = "i2c0_sda_z0";
+       function = "i2c0";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     i2c0_sck_z1_pins: i2c0-sck-z1 {
+      mux {
+       groups = "i2c0_sck_z1";
+       function = "i2c0";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     i2c0_sda_z7_pins: i2c0-sda-z7 {
+      mux {
+       groups = "i2c0_sda_z7";
+       function = "i2c0";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     i2c0_sda_z8_pins: i2c0-sda-z8 {
+      mux {
+       groups = "i2c0_sda_z8";
+       function = "i2c0";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     i2c1_sda_x_pins: i2c1-sda-x {
+      mux {
+       groups = "i2c1_sda_x";
+       function = "i2c1";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     i2c1_sck_x_pins: i2c1-sck-x {
+      mux {
+       groups = "i2c1_sck_x";
+       function = "i2c1";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     i2c1_sda_h2_pins: i2c1-sda-h2 {
+      mux {
+       groups = "i2c1_sda_h2";
+       function = "i2c1";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     i2c1_sck_h3_pins: i2c1-sck-h3 {
+      mux {
+       groups = "i2c1_sck_h3";
+       function = "i2c1";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     i2c1_sda_h6_pins: i2c1-sda-h6 {
+      mux {
+       groups = "i2c1_sda_h6";
+       function = "i2c1";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     i2c1_sck_h7_pins: i2c1-sck-h7 {
+      mux {
+       groups = "i2c1_sck_h7";
+       function = "i2c1";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     i2c2_sda_x_pins: i2c2-sda-x {
+      mux {
+       groups = "i2c2_sda_x";
+       function = "i2c2";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     i2c2_sck_x_pins: i2c2-sck-x {
+      mux {
+       groups = "i2c2_sck_x";
+       function = "i2c2";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     i2c2_sda_z_pins: i2c2-sda-z {
+      mux {
+       groups = "i2c2_sda_z";
+       function = "i2c2";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     i2c2_sck_z_pins: i2c2-sck-z {
+      mux {
+       groups = "i2c2_sck_z";
+       function = "i2c2";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     i2c3_sda_h_pins: i2c3-sda-h {
+      mux {
+       groups = "i2c3_sda_h";
+       function = "i2c3";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     i2c3_sck_h_pins: i2c3-sck-h {
+      mux {
+       groups = "i2c3_sck_h";
+       function = "i2c3";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     i2c3_sda_a_pins: i2c3-sda-a {
+      mux {
+       groups = "i2c3_sda_a";
+       function = "i2c3";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     i2c3_sck_a_pins: i2c3-sck-a {
+      mux {
+       groups = "i2c3_sck_a";
+       function = "i2c3";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     mclk0_a_pins: mclk0-a {
+      mux {
+       groups = "mclk0_a";
+       function = "mclk0";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     mclk1_a_pins: mclk1-a {
+      mux {
+       groups = "mclk1_a";
+       function = "mclk1";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     mclk1_x_pins: mclk1-x {
+      mux {
+       groups = "mclk1_x";
+       function = "mclk1";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     mclk1_z_pins: mclk1-z {
+      mux {
+       groups = "mclk1_z";
+       function = "mclk1";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     nor_pins: nor {
+      mux {
+       groups = "nor_d",
+              "nor_q",
+              "nor_c",
+              "nor_cs";
+       function = "nor";
+       bias-disable;
+      };
+     };
+
+     pdm_din0_a_pins: pdm-din0-a {
+      mux {
+       groups = "pdm_din0_a";
+       function = "pdm";
+       bias-disable;
+      };
+     };
+
+     pdm_din0_c_pins: pdm-din0-c {
+      mux {
+       groups = "pdm_din0_c";
+       function = "pdm";
+       bias-disable;
+      };
+     };
+
+     pdm_din0_x_pins: pdm-din0-x {
+      mux {
+       groups = "pdm_din0_x";
+       function = "pdm";
+       bias-disable;
+      };
+     };
+
+     pdm_din0_z_pins: pdm-din0-z {
+      mux {
+       groups = "pdm_din0_z";
+       function = "pdm";
+       bias-disable;
+      };
+     };
+
+     pdm_din1_a_pins: pdm-din1-a {
+      mux {
+       groups = "pdm_din1_a";
+       function = "pdm";
+       bias-disable;
+      };
+     };
+
+     pdm_din1_c_pins: pdm-din1-c {
+      mux {
+       groups = "pdm_din1_c";
+       function = "pdm";
+       bias-disable;
+      };
+     };
+
+     pdm_din1_x_pins: pdm-din1-x {
+      mux {
+       groups = "pdm_din1_x";
+       function = "pdm";
+       bias-disable;
+      };
+     };
+
+     pdm_din1_z_pins: pdm-din1-z {
+      mux {
+       groups = "pdm_din1_z";
+       function = "pdm";
+       bias-disable;
+      };
+     };
+
+     pdm_din2_a_pins: pdm-din2-a {
+      mux {
+       groups = "pdm_din2_a";
+       function = "pdm";
+       bias-disable;
+      };
+     };
+
+     pdm_din2_c_pins: pdm-din2-c {
+      mux {
+       groups = "pdm_din2_c";
+       function = "pdm";
+       bias-disable;
+      };
+     };
+
+     pdm_din2_x_pins: pdm-din2-x {
+      mux {
+       groups = "pdm_din2_x";
+       function = "pdm";
+       bias-disable;
+      };
+     };
+
+     pdm_din2_z_pins: pdm-din2-z {
+      mux {
+       groups = "pdm_din2_z";
+       function = "pdm";
+       bias-disable;
+      };
+     };
+
+     pdm_din3_a_pins: pdm-din3-a {
+      mux {
+       groups = "pdm_din3_a";
+       function = "pdm";
+       bias-disable;
+      };
+     };
+
+     pdm_din3_c_pins: pdm-din3-c {
+      mux {
+       groups = "pdm_din3_c";
+       function = "pdm";
+       bias-disable;
+      };
+     };
+
+     pdm_din3_x_pins: pdm-din3-x {
+      mux {
+       groups = "pdm_din3_x";
+       function = "pdm";
+       bias-disable;
+      };
+     };
+
+     pdm_din3_z_pins: pdm-din3-z {
+      mux {
+       groups = "pdm_din3_z";
+       function = "pdm";
+       bias-disable;
+      };
+     };
+
+     pdm_dclk_a_pins: pdm-dclk-a {
+      mux {
+       groups = "pdm_dclk_a";
+       function = "pdm";
+       bias-disable;
+       drive-strength-microamp = <500>;
+      };
+     };
+
+     pdm_dclk_c_pins: pdm-dclk-c {
+      mux {
+       groups = "pdm_dclk_c";
+       function = "pdm";
+       bias-disable;
+       drive-strength-microamp = <500>;
+      };
+     };
+
+     pdm_dclk_x_pins: pdm-dclk-x {
+      mux {
+       groups = "pdm_dclk_x";
+       function = "pdm";
+       bias-disable;
+       drive-strength-microamp = <500>;
+      };
+     };
+
+     pdm_dclk_z_pins: pdm-dclk-z {
+      mux {
+       groups = "pdm_dclk_z";
+       function = "pdm";
+       bias-disable;
+       drive-strength-microamp = <500>;
+      };
+     };
+
+     pwm_a_pins: pwm-a {
+      mux {
+       groups = "pwm_a";
+       function = "pwm_a";
+       bias-disable;
+      };
+     };
+
+     pwm_b_x7_pins: pwm-b-x7 {
+      mux {
+       groups = "pwm_b_x7";
+       function = "pwm_b";
+       bias-disable;
+      };
+     };
+
+     pwm_b_x19_pins: pwm-b-x19 {
+      mux {
+       groups = "pwm_b_x19";
+       function = "pwm_b";
+       bias-disable;
+      };
+     };
+
+     pwm_c_c_pins: pwm-c-c {
+      mux {
+       groups = "pwm_c_c";
+       function = "pwm_c";
+       bias-disable;
+      };
+     };
+
+     pwm_c_x5_pins: pwm-c-x5 {
+      mux {
+       groups = "pwm_c_x5";
+       function = "pwm_c";
+       bias-disable;
+      };
+     };
+
+     pwm_c_x8_pins: pwm-c-x8 {
+      mux {
+       groups = "pwm_c_x8";
+       function = "pwm_c";
+       bias-disable;
+      };
+     };
+
+     pwm_d_x3_pins: pwm-d-x3 {
+      mux {
+       groups = "pwm_d_x3";
+       function = "pwm_d";
+       bias-disable;
+      };
+     };
+
+     pwm_d_x6_pins: pwm-d-x6 {
+      mux {
+       groups = "pwm_d_x6";
+       function = "pwm_d";
+       bias-disable;
+      };
+     };
+
+     pwm_e_pins: pwm-e {
+      mux {
+       groups = "pwm_e";
+       function = "pwm_e";
+       bias-disable;
+      };
+     };
+
+     pwm_f_z_pins: pwm-f-z {
+      mux {
+       groups = "pwm_f_z";
+       function = "pwm_f";
+       bias-disable;
+      };
+     };
+
+     pwm_f_a_pins: pwm-f-a {
+      mux {
+       groups = "pwm_f_a";
+       function = "pwm_f";
+       bias-disable;
+      };
+     };
+
+     pwm_f_x_pins: pwm-f-x {
+      mux {
+       groups = "pwm_f_x";
+       function = "pwm_f";
+       bias-disable;
+      };
+     };
+
+     pwm_f_h_pins: pwm-f-h {
+      mux {
+       groups = "pwm_f_h";
+       function = "pwm_f";
+       bias-disable;
+      };
+     };
+
+     sdcard_c_pins: sdcard_c {
+      mux-0 {
+       groups = "sdcard_d0_c",
+         "sdcard_d1_c",
+         "sdcard_d2_c",
+         "sdcard_d3_c",
+         "sdcard_cmd_c";
+       function = "sdcard";
+       bias-pull-up;
+       drive-strength-microamp = <4000>;
+      };
+
+      mux-1 {
+       groups = "sdcard_clk_c";
+       function = "sdcard";
+       bias-disable;
+       drive-strength-microamp = <4000>;
+      };
+     };
+
+     sdcard_clk_gate_c_pins: sdcard_clk_gate_c {
+      mux {
+       groups = "GPIOC_4";
+       function = "gpio_periphs";
+       bias-pull-down;
+       drive-strength-microamp = <4000>;
+      };
+     };
+
+     sdcard_z_pins: sdcard_z {
+      mux-0 {
+       groups = "sdcard_d0_z",
+         "sdcard_d1_z",
+         "sdcard_d2_z",
+         "sdcard_d3_z",
+         "sdcard_cmd_z";
+       function = "sdcard";
+       bias-pull-up;
+       drive-strength-microamp = <4000>;
+      };
+
+      mux-1 {
+       groups = "sdcard_clk_z";
+       function = "sdcard";
+       bias-disable;
+       drive-strength-microamp = <4000>;
+      };
+     };
+
+     sdcard_clk_gate_z_pins: sdcard_clk_gate_z {
+      mux {
+       groups = "GPIOZ_6";
+       function = "gpio_periphs";
+       bias-pull-down;
+       drive-strength-microamp = <4000>;
+      };
+     };
+
+     sdio_pins: sdio {
+      mux {
+       groups = "sdio_d0",
+         "sdio_d1",
+         "sdio_d2",
+         "sdio_d3",
+         "sdio_clk",
+         "sdio_cmd";
+       function = "sdio";
+       bias-disable;
+       drive-strength-microamp = <4000>;
+      };
+     };
+
+     sdio_clk_gate_pins: sdio_clk_gate {
+      mux {
+       groups = "GPIOX_4";
+       function = "gpio_periphs";
+       bias-pull-down;
+       drive-strength-microamp = <4000>;
+      };
+     };
+
+     spdif_in_a10_pins: spdif-in-a10 {
+      mux {
+       groups = "spdif_in_a10";
+       function = "spdif_in";
+       bias-disable;
+      };
+     };
+
+     spdif_in_a12_pins: spdif-in-a12 {
+      mux {
+       groups = "spdif_in_a12";
+       function = "spdif_in";
+       bias-disable;
+      };
+     };
+
+     spdif_in_h_pins: spdif-in-h {
+      mux {
+       groups = "spdif_in_h";
+       function = "spdif_in";
+       bias-disable;
+      };
+     };
+
+     spdif_out_h_pins: spdif-out-h {
+      mux {
+       groups = "spdif_out_h";
+       function = "spdif_out";
+       drive-strength-microamp = <500>;
+       bias-disable;
+      };
+     };
+
+     spdif_out_a11_pins: spdif-out-a11 {
+      mux {
+       groups = "spdif_out_a11";
+       function = "spdif_out";
+       drive-strength-microamp = <500>;
+       bias-disable;
+      };
+     };
+
+     spdif_out_a13_pins: spdif-out-a13 {
+      mux {
+       groups = "spdif_out_a13";
+       function = "spdif_out";
+       drive-strength-microamp = <500>;
+       bias-disable;
+      };
+     };
+
+     spicc0_x_pins: spicc0-x {
+      mux {
+       groups = "spi0_mosi_x",
+              "spi0_miso_x",
+              "spi0_clk_x";
+       function = "spi0";
+       drive-strength-microamp = <4000>;
+       bias-disable;
+      };
+     };
+
+     spicc0_ss0_x_pins: spicc0-ss0-x {
+      mux {
+       groups = "spi0_ss0_x";
+       function = "spi0";
+       drive-strength-microamp = <4000>;
+       bias-disable;
+      };
+     };
+
+     spicc0_c_pins: spicc0-c {
+      mux {
+       groups = "spi0_mosi_c",
+              "spi0_miso_c",
+              "spi0_ss0_c",
+              "spi0_clk_c";
+       function = "spi0";
+       drive-strength-microamp = <4000>;
+       bias-disable;
+      };
+     };
+
+     spicc1_pins: spicc1 {
+      mux {
+       groups = "spi1_mosi",
+              "spi1_miso",
+              "spi1_clk";
+       function = "spi1";
+       drive-strength-microamp = <4000>;
+      };
+     };
+
+     spicc1_ss0_pins: spicc1-ss0 {
+      mux {
+       groups = "spi1_ss0";
+       function = "spi1";
+       drive-strength-microamp = <4000>;
+       bias-disable;
+      };
+     };
+
+     tdm_a_din0_pins: tdm-a-din0 {
+      mux {
+       groups = "tdm_a_din0";
+       function = "tdm_a";
+       bias-disable;
+      };
+     };
+
+
+     tdm_a_din1_pins: tdm-a-din1 {
+      mux {
+       groups = "tdm_a_din1";
+       function = "tdm_a";
+       bias-disable;
+      };
+     };
+
+     tdm_a_dout0_pins: tdm-a-dout0 {
+      mux {
+       groups = "tdm_a_dout0";
+       function = "tdm_a";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     tdm_a_dout1_pins: tdm-a-dout1 {
+      mux {
+       groups = "tdm_a_dout1";
+       function = "tdm_a";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     tdm_a_fs_pins: tdm-a-fs {
+      mux {
+       groups = "tdm_a_fs";
+       function = "tdm_a";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     tdm_a_sclk_pins: tdm-a-sclk {
+      mux {
+       groups = "tdm_a_sclk";
+       function = "tdm_a";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     tdm_a_slv_fs_pins: tdm-a-slv-fs {
+      mux {
+       groups = "tdm_a_slv_fs";
+       function = "tdm_a";
+       bias-disable;
+      };
+     };
+
+
+     tdm_a_slv_sclk_pins: tdm-a-slv-sclk {
+      mux {
+       groups = "tdm_a_slv_sclk";
+       function = "tdm_a";
+       bias-disable;
+      };
+     };
+
+     tdm_b_din0_pins: tdm-b-din0 {
+      mux {
+       groups = "tdm_b_din0";
+       function = "tdm_b";
+       bias-disable;
+      };
+     };
+
+     tdm_b_din1_pins: tdm-b-din1 {
+      mux {
+       groups = "tdm_b_din1";
+       function = "tdm_b";
+       bias-disable;
+      };
+     };
+
+     tdm_b_din2_pins: tdm-b-din2 {
+      mux {
+       groups = "tdm_b_din2";
+       function = "tdm_b";
+       bias-disable;
+      };
+     };
+
+     tdm_b_din3_a_pins: tdm-b-din3-a {
+      mux {
+       groups = "tdm_b_din3_a";
+       function = "tdm_b";
+       bias-disable;
+      };
+     };
+
+     tdm_b_din3_h_pins: tdm-b-din3-h {
+      mux {
+       groups = "tdm_b_din3_h";
+       function = "tdm_b";
+       bias-disable;
+      };
+     };
+
+     tdm_b_dout0_pins: tdm-b-dout0 {
+      mux {
+       groups = "tdm_b_dout0";
+       function = "tdm_b";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     tdm_b_dout1_pins: tdm-b-dout1 {
+      mux {
+       groups = "tdm_b_dout1";
+       function = "tdm_b";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     tdm_b_dout2_pins: tdm-b-dout2 {
+      mux {
+       groups = "tdm_b_dout2";
+       function = "tdm_b";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     tdm_b_dout3_a_pins: tdm-b-dout3-a {
+      mux {
+       groups = "tdm_b_dout3_a";
+       function = "tdm_b";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     tdm_b_dout3_h_pins: tdm-b-dout3-h {
+      mux {
+       groups = "tdm_b_dout3_h";
+       function = "tdm_b";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     tdm_b_fs_pins: tdm-b-fs {
+      mux {
+       groups = "tdm_b_fs";
+       function = "tdm_b";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     tdm_b_sclk_pins: tdm-b-sclk {
+      mux {
+       groups = "tdm_b_sclk";
+       function = "tdm_b";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     tdm_b_slv_fs_pins: tdm-b-slv-fs {
+      mux {
+       groups = "tdm_b_slv_fs";
+       function = "tdm_b";
+       bias-disable;
+      };
+     };
+
+     tdm_b_slv_sclk_pins: tdm-b-slv-sclk {
+      mux {
+       groups = "tdm_b_slv_sclk";
+       function = "tdm_b";
+       bias-disable;
+      };
+     };
+
+     tdm_c_din0_a_pins: tdm-c-din0-a {
+      mux {
+       groups = "tdm_c_din0_a";
+       function = "tdm_c";
+       bias-disable;
+      };
+     };
+
+     tdm_c_din0_z_pins: tdm-c-din0-z {
+      mux {
+       groups = "tdm_c_din0_z";
+       function = "tdm_c";
+       bias-disable;
+      };
+     };
+
+     tdm_c_din1_a_pins: tdm-c-din1-a {
+      mux {
+       groups = "tdm_c_din1_a";
+       function = "tdm_c";
+       bias-disable;
+      };
+     };
+
+     tdm_c_din1_z_pins: tdm-c-din1-z {
+      mux {
+       groups = "tdm_c_din1_z";
+       function = "tdm_c";
+       bias-disable;
+      };
+     };
+
+     tdm_c_din2_a_pins: tdm-c-din2-a {
+      mux {
+       groups = "tdm_c_din2_a";
+       function = "tdm_c";
+       bias-disable;
+      };
+     };
+
+     eth_leds_pins: eth-leds {
+      mux {
+       groups = "eth_link_led",
+         "eth_act_led";
+       function = "eth";
+       bias-disable;
+      };
+     };
+
+     eth_pins: eth {
+      mux {
+       groups = "eth_mdio",
+         "eth_mdc",
+         "eth_rgmii_rx_clk",
+         "eth_rx_dv",
+         "eth_rxd0",
+         "eth_rxd1",
+         "eth_txen",
+         "eth_txd0",
+         "eth_txd1";
+       function = "eth";
+       drive-strength-microamp = <4000>;
+       bias-disable;
+      };
+     };
+
+     eth_rgmii_pins: eth-rgmii {
+      mux {
+       groups = "eth_rxd2_rgmii",
+         "eth_rxd3_rgmii",
+         "eth_rgmii_tx_clk",
+         "eth_txd2_rgmii",
+         "eth_txd3_rgmii";
+       function = "eth";
+       drive-strength-microamp = <4000>;
+       bias-disable;
+      };
+     };
+
+     tdm_c_din2_z_pins: tdm-c-din2-z {
+      mux {
+       groups = "tdm_c_din2_z";
+       function = "tdm_c";
+       bias-disable;
+      };
+     };
+
+     tdm_c_din3_a_pins: tdm-c-din3-a {
+      mux {
+       groups = "tdm_c_din3_a";
+       function = "tdm_c";
+       bias-disable;
+      };
+     };
+
+     tdm_c_din3_z_pins: tdm-c-din3-z {
+      mux {
+       groups = "tdm_c_din3_z";
+       function = "tdm_c";
+       bias-disable;
+      };
+     };
+
+     tdm_c_dout0_a_pins: tdm-c-dout0-a {
+      mux {
+       groups = "tdm_c_dout0_a";
+       function = "tdm_c";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     tdm_c_dout0_z_pins: tdm-c-dout0-z {
+      mux {
+       groups = "tdm_c_dout0_z";
+       function = "tdm_c";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     tdm_c_dout1_a_pins: tdm-c-dout1-a {
+      mux {
+       groups = "tdm_c_dout1_a";
+       function = "tdm_c";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     tdm_c_dout1_z_pins: tdm-c-dout1-z {
+      mux {
+       groups = "tdm_c_dout1_z";
+       function = "tdm_c";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     tdm_c_dout2_a_pins: tdm-c-dout2-a {
+      mux {
+       groups = "tdm_c_dout2_a";
+       function = "tdm_c";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     tdm_c_dout2_z_pins: tdm-c-dout2-z {
+      mux {
+       groups = "tdm_c_dout2_z";
+       function = "tdm_c";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     tdm_c_dout3_a_pins: tdm-c-dout3-a {
+      mux {
+       groups = "tdm_c_dout3_a";
+       function = "tdm_c";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     tdm_c_dout3_z_pins: tdm-c-dout3-z {
+      mux {
+       groups = "tdm_c_dout3_z";
+       function = "tdm_c";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     tdm_c_fs_a_pins: tdm-c-fs-a {
+      mux {
+       groups = "tdm_c_fs_a";
+       function = "tdm_c";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     tdm_c_fs_z_pins: tdm-c-fs-z {
+      mux {
+       groups = "tdm_c_fs_z";
+       function = "tdm_c";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     tdm_c_sclk_a_pins: tdm-c-sclk-a {
+      mux {
+       groups = "tdm_c_sclk_a";
+       function = "tdm_c";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     tdm_c_sclk_z_pins: tdm-c-sclk-z {
+      mux {
+       groups = "tdm_c_sclk_z";
+       function = "tdm_c";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     tdm_c_slv_fs_a_pins: tdm-c-slv-fs-a {
+      mux {
+       groups = "tdm_c_slv_fs_a";
+       function = "tdm_c";
+       bias-disable;
+      };
+     };
+
+     tdm_c_slv_fs_z_pins: tdm-c-slv-fs-z {
+      mux {
+       groups = "tdm_c_slv_fs_z";
+       function = "tdm_c";
+       bias-disable;
+      };
+     };
+
+     tdm_c_slv_sclk_a_pins: tdm-c-slv-sclk-a {
+      mux {
+       groups = "tdm_c_slv_sclk_a";
+       function = "tdm_c";
+       bias-disable;
+      };
+     };
+
+     tdm_c_slv_sclk_z_pins: tdm-c-slv-sclk-z {
+      mux {
+       groups = "tdm_c_slv_sclk_z";
+       function = "tdm_c";
+       bias-disable;
+      };
+     };
+
+     uart_a_pins: uart-a {
+      mux {
+       groups = "uart_a_tx",
+         "uart_a_rx";
+       function = "uart_a";
+       bias-disable;
+      };
+     };
+
+     uart_a_cts_rts_pins: uart-a-cts-rts {
+      mux {
+       groups = "uart_a_cts",
+         "uart_a_rts";
+       function = "uart_a";
+       bias-disable;
+      };
+     };
+
+     uart_b_pins: uart-b {
+      mux {
+       groups = "uart_b_tx",
+         "uart_b_rx";
+       function = "uart_b";
+       bias-disable;
+      };
+     };
+
+     uart_c_pins: uart-c {
+      mux {
+       groups = "uart_c_tx",
+         "uart_c_rx";
+       function = "uart_c";
+       bias-disable;
+      };
+     };
+
+     uart_c_cts_rts_pins: uart-c-cts-rts {
+      mux {
+       groups = "uart_c_cts",
+         "uart_c_rts";
+       function = "uart_c";
+       bias-disable;
+      };
+     };
+    };
+   };
+
+   cpu_temp: temperature-sensor@34800 {
+    compatible = "amlogic,g12a-cpu-thermal",
+          "amlogic,g12a-thermal";
+    reg = <0x0 0x34800 0x0 0x50>;
+    interrupts = <0 35 1>;
+    clocks = <&clkc 212>;
+    #thermal-sensor-cells = <0>;
+    amlogic,ao-secure = <&sec_AO>;
+   };
+
+   ddr_temp: temperature-sensor@34c00 {
+    compatible = "amlogic,g12a-ddr-thermal",
+          "amlogic,g12a-thermal";
+    reg = <0x0 0x34c00 0x0 0x50>;
+    interrupts = <0 36 1>;
+    clocks = <&clkc 212>;
+    #thermal-sensor-cells = <0>;
+    amlogic,ao-secure = <&sec_AO>;
+   };
+
+   usb2_phy0: phy@36000 {
+    compatible = "amlogic,g12a-usb2-phy";
+    reg = <0x0 0x36000 0x0 0x2000>;
+    clocks = <&xtal>;
+    clock-names = "xtal";
+    resets = <&reset 48>;
+    reset-names = "phy";
+    #phy-cells = <0>;
+   };
+
+   dmc: bus@38000 {
+    compatible = "simple-bus";
+    reg = <0x0 0x38000 0x0 0x400>;
+    #address-cells = <2>;
+    #size-cells = <2>;
+    ranges = <0x0 0x0 0x0 0x38000 0x0 0x400>;
+
+    canvas: video-lut@48 {
+     compatible = "amlogic,canvas";
+     reg = <0x0 0x48 0x0 0x14>;
+    };
+   };
+
+   usb2_phy1: phy@3a000 {
+    compatible = "amlogic,g12a-usb2-phy";
+    reg = <0x0 0x3a000 0x0 0x2000>;
+    clocks = <&xtal>;
+    clock-names = "xtal";
+    resets = <&reset 49>;
+    reset-names = "phy";
+    #phy-cells = <0>;
+   };
+
+   hiu: bus@3c000 {
+    compatible = "simple-bus";
+    reg = <0x0 0x3c000 0x0 0x1400>;
+    #address-cells = <2>;
+    #size-cells = <2>;
+    ranges = <0x0 0x0 0x0 0x3c000 0x0 0x1400>;
+
+    hhi: system-controller@0 {
+     compatible = "amlogic,meson-gx-hhi-sysctrl",
+           "simple-mfd", "syscon";
+     reg = <0 0 0 0x400>;
+
+     clkc: clock-controller {
+      compatible = "amlogic,g12a-clkc";
+      #clock-cells = <1>;
+      clocks = <&xtal>;
+      clock-names = "xtal";
+     };
+
+     pwrc: power-controller {
+      compatible = "amlogic,meson-g12a-pwrc";
+      #power-domain-cells = <1>;
+      amlogic,ao-sysctrl = <&rti>;
+      resets = <&reset 5>,
+        <&reset 10>,
+        <&reset 13>,
+        <&reset 37>,
+        <&reset 133>,
+        <&reset 134>,
+        <&reset 135>,
+        <&reset 137>,
+        <&reset 140>,
+        <&reset 141>,
+        <&reset 231>;
+      reset-names = "viu", "venc", "vcbus", "bt656",
+             "rdma", "venci", "vencp", "vdac",
+             "vdi6", "vencl", "vid_lock";
+      clocks = <&clkc 116>,
+        <&clkc 124>;
+      clock-names = "vpu", "vapb";
+
+
+
+
+
+
+      assigned-clocks = <&clkc 110>,
+          <&clkc 112>,
+          <&clkc 116>,
+          <&clkc 117>,
+          <&clkc 119>,
+          <&clkc 123>;
+      assigned-clock-parents = <&clkc 3>,
+          <0>,
+          <&clkc 112>,
+          <&clkc 4>,
+          <0>,
+          <&clkc 119>;
+      assigned-clock-rates = <0>,
+               <666666666>,
+               <0>,
+               <0>,
+               <250000000>,
+               <0>;
+     };
+    };
+   };
+
+   usb3_pcie_phy: phy@46000 {
+    compatible = "amlogic,g12a-usb3-pcie-phy";
+    reg = <0x0 0x46000 0x0 0x2000>;
+    clocks = <&clkc 201>;
+    clock-names = "ref_clk";
+    resets = <&reset 14>;
+    reset-names = "phy";
+    assigned-clocks = <&clkc 201>;
+    assigned-clock-rates = <100000000>;
+    #phy-cells = <1>;
+   };
+
+   eth_phy: mdio-multiplexer@4c000 {
+    compatible = "amlogic,g12a-mdio-mux";
+    reg = <0x0 0x4c000 0x0 0xa4>;
+    clocks = <&clkc 19>,
+      <&xtal>,
+      <&clkc 177>;
+    clock-names = "pclk", "clkin0", "clkin1";
+    mdio-parent-bus = <&mdio0>;
+    #address-cells = <1>;
+    #size-cells = <0>;
+
+    ext_mdio: mdio@0 {
+     reg = <0>;
+     #address-cells = <1>;
+     #size-cells = <0>;
+    };
+
+    int_mdio: mdio@1 {
+     reg = <1>;
+     #address-cells = <1>;
+     #size-cells = <0>;
+
+     internal_ephy: ethernet_phy@8 {
+      compatible = "ethernet-phy-id0180.3301",
+            "ethernet-phy-ieee802.3-c22";
+      interrupts = <0 9 4>;
+      reg = <8>;
+      max-speed = <100>;
+     };
+    };
+   };
+  };
+
+  aobus: bus@ff800000 {
+   compatible = "simple-bus";
+   reg = <0x0 0xff800000 0x0 0x100000>;
+   #address-cells = <2>;
+   #size-cells = <2>;
+   ranges = <0x0 0x0 0x0 0xff800000 0x0 0x100000>;
+
+   rti: sys-ctrl@0 {
+    compatible = "amlogic,meson-gx-ao-sysctrl",
+          "simple-mfd", "syscon";
+    reg = <0x0 0x0 0x0 0x100>;
+    #address-cells = <2>;
+    #size-cells = <2>;
+    ranges = <0x0 0x0 0x0 0x0 0x0 0x100>;
+
+    clkc_AO: clock-controller {
+     compatible = "amlogic,meson-g12a-aoclkc";
+     #clock-cells = <1>;
+     #reset-cells = <1>;
+     clocks = <&xtal>, <&clkc 10>;
+     clock-names = "xtal", "mpeg-clk";
+    };
+
+    ao_pinctrl: pinctrl@14 {
+     compatible = "amlogic,meson-g12a-aobus-pinctrl";
+     #address-cells = <2>;
+     #size-cells = <2>;
+     ranges;
+
+     gpio_ao: bank@14 {
+      reg = <0x0 0x14 0x0 0x8>,
+            <0x0 0x1c 0x0 0x8>,
+            <0x0 0x24 0x0 0x14>;
+      reg-names = "mux",
+           "ds",
+           "gpio";
+      gpio-controller;
+      #gpio-cells = <2>;
+      gpio-ranges = <&ao_pinctrl 0 0 15>;
+     };
+
+     i2c_ao_sck_pins: i2c_ao_sck_pins {
+      mux {
+       groups = "i2c_ao_sck";
+       function = "i2c_ao";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     i2c_ao_sda_pins: i2c_ao_sda {
+      mux {
+       groups = "i2c_ao_sda";
+       function = "i2c_ao";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     i2c_ao_sck_e_pins: i2c_ao_sck_e {
+      mux {
+       groups = "i2c_ao_sck_e";
+       function = "i2c_ao";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     i2c_ao_sda_e_pins: i2c_ao_sda_e {
+      mux {
+       groups = "i2c_ao_sda_e";
+       function = "i2c_ao";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     mclk0_ao_pins: mclk0-ao {
+      mux {
+       groups = "mclk0_ao";
+       function = "mclk0_ao";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     tdm_ao_b_din0_pins: tdm-ao-b-din0 {
+      mux {
+       groups = "tdm_ao_b_din0";
+       function = "tdm_ao_b";
+       bias-disable;
+      };
+     };
+
+     spdif_ao_out_pins: spdif-ao-out {
+      mux {
+       groups = "spdif_ao_out";
+       function = "spdif_ao_out";
+       drive-strength-microamp = <500>;
+       bias-disable;
+      };
+     };
+
+     tdm_ao_b_din1_pins: tdm-ao-b-din1 {
+      mux {
+       groups = "tdm_ao_b_din1";
+       function = "tdm_ao_b";
+       bias-disable;
+      };
+     };
+
+     tdm_ao_b_din2_pins: tdm-ao-b-din2 {
+      mux {
+       groups = "tdm_ao_b_din2";
+       function = "tdm_ao_b";
+       bias-disable;
+      };
+     };
+
+     tdm_ao_b_dout0_pins: tdm-ao-b-dout0 {
+      mux {
+       groups = "tdm_ao_b_dout0";
+       function = "tdm_ao_b";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     tdm_ao_b_dout1_pins: tdm-ao-b-dout1 {
+      mux {
+       groups = "tdm_ao_b_dout1";
+       function = "tdm_ao_b";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     tdm_ao_b_dout2_pins: tdm-ao-b-dout2 {
+      mux {
+       groups = "tdm_ao_b_dout2";
+       function = "tdm_ao_b";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     tdm_ao_b_fs_pins: tdm-ao-b-fs {
+      mux {
+       groups = "tdm_ao_b_fs";
+       function = "tdm_ao_b";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     tdm_ao_b_sclk_pins: tdm-ao-b-sclk {
+      mux {
+       groups = "tdm_ao_b_sclk";
+       function = "tdm_ao_b";
+       bias-disable;
+       drive-strength-microamp = <3000>;
+      };
+     };
+
+     tdm_ao_b_slv_fs_pins: tdm-ao-b-slv-fs {
+      mux {
+       groups = "tdm_ao_b_slv_fs";
+       function = "tdm_ao_b";
+       bias-disable;
+      };
+     };
+
+     tdm_ao_b_slv_sclk_pins: tdm-ao-b-slv-sclk {
+      mux {
+       groups = "tdm_ao_b_slv_sclk";
+       function = "tdm_ao_b";
+       bias-disable;
+      };
+     };
+
+     uart_ao_a_pins: uart-a-ao {
+      mux {
+       groups = "uart_ao_a_tx",
+         "uart_ao_a_rx";
+       function = "uart_ao_a";
+       bias-disable;
+      };
+     };
+
+     uart_ao_a_cts_rts_pins: uart-ao-a-cts-rts {
+      mux {
+       groups = "uart_ao_a_cts",
+         "uart_ao_a_rts";
+       function = "uart_ao_a";
+       bias-disable;
+      };
+     };
+
+     uart_ao_b_2_3_pins: uart-ao-b-2-3 {
+      mux {
+       groups = "uart_ao_b_tx_2",
+         "uart_ao_b_rx_3";
+       function = "uart_ao_b";
+       bias-disable;
+      };
+     };
+
+     uart_ao_b_8_9_pins: uart-ao-b-8-9 {
+      mux {
+       groups = "uart_ao_b_tx_8",
+         "uart_ao_b_rx_9";
+       function = "uart_ao_b";
+       bias-disable;
+      };
+     };
+
+     uart_ao_b_cts_rts_pins: uart-ao-b-cts-rts {
+      mux {
+       groups = "uart_ao_b_cts",
+         "uart_ao_b_rts";
+       function = "uart_ao_b";
+       bias-disable;
+      };
+     };
+
+     pwm_a_e_pins: pwm-a-e {
+      mux {
+       groups = "pwm_a_e";
+       function = "pwm_a_e";
+       bias-disable;
+      };
+     };
+
+     pwm_ao_a_pins: pwm-ao-a {
+      mux {
+       groups = "pwm_ao_a";
+       function = "pwm_ao_a";
+       bias-disable;
+      };
+     };
+
+     pwm_ao_b_pins: pwm-ao-b {
+      mux {
+       groups = "pwm_ao_b";
+       function = "pwm_ao_b";
+       bias-disable;
+      };
+     };
+
+     pwm_ao_c_4_pins: pwm-ao-c-4 {
+      mux {
+       groups = "pwm_ao_c_4";
+       function = "pwm_ao_c";
+       bias-disable;
+      };
+     };
+
+     pwm_ao_c_6_pins: pwm-ao-c-6 {
+      mux {
+       groups = "pwm_ao_c_6";
+       function = "pwm_ao_c";
+       bias-disable;
+      };
+     };
+
+     pwm_ao_d_5_pins: pwm-ao-d-5 {
+      mux {
+       groups = "pwm_ao_d_5";
+       function = "pwm_ao_d";
+       bias-disable;
+      };
+     };
+
+     pwm_ao_d_10_pins: pwm-ao-d-10 {
+      mux {
+       groups = "pwm_ao_d_10";
+       function = "pwm_ao_d";
+       bias-disable;
+      };
+     };
+
+     pwm_ao_d_e_pins: pwm-ao-d-e {
+      mux {
+       groups = "pwm_ao_d_e";
+       function = "pwm_ao_d";
+      };
+     };
+
+     remote_input_ao_pins: remote-input-ao {
+      mux {
+       groups = "remote_ao_input";
+       function = "remote_ao_input";
+       bias-disable;
+      };
+     };
+    };
+   };
+
+   vrtc: rtc@a8 {
+    compatible = "amlogic,meson-vrtc";
+    reg = <0x0 0x000a8 0x0 0x4>;
+   };
+
+   cec_AO: cec@100 {
+    compatible = "amlogic,meson-gx-ao-cec";
+    reg = <0x0 0x00100 0x0 0x14>;
+    interrupts = <0 199 1>;
+    clocks = <&clkc_AO 27>;
+    clock-names = "core";
+    status = "disabled";
+   };
+
+   sec_AO: ao-secure@140 {
+    compatible = "amlogic,meson-gx-ao-secure", "syscon";
+    reg = <0x0 0x140 0x0 0x140>;
+    amlogic,has-chip-id;
+   };
+
+   cecb_AO: cec@280 {
+    compatible = "amlogic,meson-g12a-ao-cec";
+    reg = <0x0 0x00280 0x0 0x1c>;
+    interrupts = <0 203 1>;
+    clocks = <&clkc_AO 19>;
+    clock-names = "oscin";
+    status = "disabled";
+   };
+
+   pwm_AO_cd: pwm@2000 {
+    compatible = "amlogic,meson-g12a-ao-pwm-cd";
+    reg = <0x0 0x2000 0x0 0x20>;
+    #pwm-cells = <3>;
+    status = "disabled";
+   };
+
+   uart_AO: serial@3000 {
+    compatible = "amlogic,meson-gx-uart",
+          "amlogic,meson-ao-uart";
+    reg = <0x0 0x3000 0x0 0x18>;
+    interrupts = <0 193 1>;
+    clocks = <&xtal>, <&clkc_AO 4>, <&xtal>;
+    clock-names = "xtal", "pclk", "baud";
+    status = "disabled";
+   };
+
+   uart_AO_B: serial@4000 {
+    compatible = "amlogic,meson-gx-uart",
+          "amlogic,meson-ao-uart";
+    reg = <0x0 0x4000 0x0 0x18>;
+    interrupts = <0 197 1>;
+    clocks = <&xtal>, <&clkc_AO 6>, <&xtal>;
+    clock-names = "xtal", "pclk", "baud";
+    status = "disabled";
+   };
+
+   i2c_AO: i2c@5000 {
+    compatible = "amlogic,meson-axg-i2c";
+    status = "disabled";
+    reg = <0x0 0x05000 0x0 0x20>;
+    interrupts = <0 195 1>;
+    #address-cells = <1>;
+    #size-cells = <0>;
+    clocks = <&clkc 24>;
+   };
+
+   pwm_AO_ab: pwm@7000 {
+    compatible = "amlogic,meson-g12a-ao-pwm-ab";
+    reg = <0x0 0x7000 0x0 0x20>;
+    #pwm-cells = <3>;
+    status = "disabled";
+   };
+
+   ir: ir@8000 {
+    compatible = "amlogic,meson-gxbb-ir";
+    reg = <0x0 0x8000 0x0 0x20>;
+    interrupts = <0 196 1>;
+    status = "disabled";
+   };
+
+   saradc: adc@9000 {
+    compatible = "amlogic,meson-g12a-saradc",
+          "amlogic,meson-saradc";
+    reg = <0x0 0x9000 0x0 0x48>;
+    #io-channel-cells = <1>;
+    interrupts = <0 200 1>;
+    clocks = <&xtal>,
+      <&clkc_AO 8>,
+      <&clkc_AO 18>,
+      <&clkc_AO 16>;
+    clock-names = "clkin", "core", "adc_clk", "adc_sel";
+    status = "disabled";
+   };
+  };
+
+  vdec: video-decoder@ff620000 {
+   compatible = "amlogic,g12a-vdec";
+   reg = <0x0 0xff620000 0x0 0x10000>,
+         <0x0 0xffd0e180 0x0 0xe4>;
+   reg-names = "dos", "esparser";
+   interrupts = <0 44 1>,
+         <0 32 1>;
+   interrupt-names = "vdec", "esparser";
+
+   amlogic,ao-sysctrl = <&rti>;
+   amlogic,canvas = <&canvas>;
+
+   clocks = <&clkc 46>,
+     <&clkc 16>,
+     <&clkc 204>,
+     <&clkc 207>,
+     <&clkc 210>;
+   clock-names = "dos_parser", "dos", "vdec_1",
+          "vdec_hevc", "vdec_hevcf";
+   resets = <&reset 40>;
+   reset-names = "esparser";
+  };
+
+  vpu: vpu@ff900000 {
+   compatible = "amlogic,meson-g12a-vpu";
+   reg = <0x0 0xff900000 0x0 0x100000>,
+         <0x0 0xff63c000 0x0 0x1000>;
+   reg-names = "vpu", "hhi";
+   interrupts = <0 3 1>;
+   #address-cells = <1>;
+   #size-cells = <0>;
+   amlogic,canvas = <&canvas>;
+
+
+   cvbs_vdac_port: port@0 {
+    reg = <0>;
+   };
+
+
+   hdmi_tx_port: port@1 {
+    reg = <1>;
+
+    hdmi_tx_out: endpoint {
+     remote-endpoint = <&hdmi_tx_in>;
+    };
+   };
+  };
+
+  gic: interrupt-controller@ffc01000 {
+   compatible = "arm,gic-400";
+   reg = <0x0 0xffc01000 0 0x1000>,
+         <0x0 0xffc02000 0 0x2000>,
+         <0x0 0xffc04000 0 0x2000>,
+         <0x0 0xffc06000 0 0x2000>;
+   interrupt-controller;
+   interrupts = <1 9
+    ((((1 << (8)) - 1) << 8) | 4)>;
+   #interrupt-cells = <3>;
+   #address-cells = <0>;
+  };
+
+  cbus: bus@ffd00000 {
+   compatible = "simple-bus";
+   reg = <0x0 0xffd00000 0x0 0x100000>;
+   #address-cells = <2>;
+   #size-cells = <2>;
+   ranges = <0x0 0x0 0x0 0xffd00000 0x0 0x100000>;
+
+   reset: reset-controller@1004 {
+    compatible = "amlogic,meson-axg-reset";
+    reg = <0x0 0x1004 0x0 0x9c>;
+    #reset-cells = <1>;
+   };
+
+   gpio_intc: interrupt-controller@f080 {
+    compatible = "amlogic,meson-g12a-gpio-intc",
+          "amlogic,meson-gpio-intc";
+    reg = <0x0 0xf080 0x0 0x10>;
+    interrupt-controller;
+    #interrupt-cells = <2>;
+    amlogic,channel-interrupts = <64 65 66 67 68 69 70 71>;
+   };
+
+   watchdog: watchdog@f0d0 {
+    compatible = "amlogic,meson-gxbb-wdt";
+    reg = <0x0 0xf0d0 0x0 0x10>;
+    clocks = <&xtal>;
+   };
+
+   spicc0: spi@13000 {
+    compatible = "amlogic,meson-g12a-spicc";
+    reg = <0x0 0x13000 0x0 0x44>;
+    interrupts = <0 81 4>;
+    clocks = <&clkc 23>,
+      <&clkc 258>;
+    clock-names = "core", "pclk";
+    #address-cells = <1>;
+    #size-cells = <0>;
+    status = "disabled";
+   };
+
+   spicc1: spi@15000 {
+    compatible = "amlogic,meson-g12a-spicc";
+    reg = <0x0 0x15000 0x0 0x44>;
+    interrupts = <0 90 4>;
+    clocks = <&clkc 29>,
+      <&clkc 261>;
+    clock-names = "core", "pclk";
+    #address-cells = <1>;
+    #size-cells = <0>;
+    status = "disabled";
+   };
+
+   spifc: spi@14000 {
+    compatible = "amlogic,meson-gxbb-spifc";
+    status = "disabled";
+    reg = <0x0 0x14000 0x0 0x80>;
+    #address-cells = <1>;
+    #size-cells = <0>;
+    clocks = <&clkc 10>;
+   };
+
+   pwm_ef: pwm@19000 {
+    compatible = "amlogic,meson-g12a-ee-pwm";
+    reg = <0x0 0x19000 0x0 0x20>;
+    #pwm-cells = <3>;
+    status = "disabled";
+   };
+
+   pwm_cd: pwm@1a000 {
+    compatible = "amlogic,meson-g12a-ee-pwm";
+    reg = <0x0 0x1a000 0x0 0x20>;
+    #pwm-cells = <3>;
+    status = "disabled";
+   };
+
+   pwm_ab: pwm@1b000 {
+    compatible = "amlogic,meson-g12a-ee-pwm";
+    reg = <0x0 0x1b000 0x0 0x20>;
+    #pwm-cells = <3>;
+    status = "disabled";
+   };
+
+   i2c3: i2c@1c000 {
+    compatible = "amlogic,meson-axg-i2c";
+    status = "disabled";
+    reg = <0x0 0x1c000 0x0 0x20>;
+    interrupts = <0 39 1>;
+    #address-cells = <1>;
+    #size-cells = <0>;
+    clocks = <&clkc 24>;
+   };
+
+   i2c2: i2c@1d000 {
+    compatible = "amlogic,meson-axg-i2c";
+    status = "disabled";
+    reg = <0x0 0x1d000 0x0 0x20>;
+    interrupts = <0 215 1>;
+    #address-cells = <1>;
+    #size-cells = <0>;
+    clocks = <&clkc 24>;
+   };
+
+   i2c1: i2c@1e000 {
+    compatible = "amlogic,meson-axg-i2c";
+    status = "disabled";
+    reg = <0x0 0x1e000 0x0 0x20>;
+    interrupts = <0 214 1>;
+    #address-cells = <1>;
+    #size-cells = <0>;
+    clocks = <&clkc 24>;
+   };
+
+   i2c0: i2c@1f000 {
+    compatible = "amlogic,meson-axg-i2c";
+    status = "disabled";
+    reg = <0x0 0x1f000 0x0 0x20>;
+    interrupts = <0 21 1>;
+    #address-cells = <1>;
+    #size-cells = <0>;
+    clocks = <&clkc 24>;
+   };
+
+   clk_msr: clock-measure@18000 {
+    compatible = "amlogic,meson-g12a-clk-measure";
+    reg = <0x0 0x18000 0x0 0x10>;
+   };
+
+   uart_C: serial@22000 {
+    compatible = "amlogic,meson-gx-uart";
+    reg = <0x0 0x22000 0x0 0x18>;
+    interrupts = <0 93 1>;
+    clocks = <&xtal>, <&clkc 57>, <&xtal>;
+    clock-names = "xtal", "pclk", "baud";
+    status = "disabled";
+   };
+
+   uart_B: serial@23000 {
+    compatible = "amlogic,meson-gx-uart";
+    reg = <0x0 0x23000 0x0 0x18>;
+    interrupts = <0 75 1>;
+    clocks = <&xtal>, <&clkc 42>, <&xtal>;
+    clock-names = "xtal", "pclk", "baud";
+    status = "disabled";
+   };
+
+   uart_A: serial@24000 {
+    compatible = "amlogic,meson-gx-uart";
+    reg = <0x0 0x24000 0x0 0x18>;
+    interrupts = <0 26 1>;
+    clocks = <&xtal>, <&clkc 28>, <&xtal>;
+    clock-names = "xtal", "pclk", "baud";
+    status = "disabled";
+    fifo-size = <128>;
+   };
+  };
+
+  sd_emmc_a: sd@ffe03000 {
+   compatible = "amlogic,meson-axg-mmc";
+   reg = <0x0 0xffe03000 0x0 0x800>;
+   interrupts = <0 189 4>;
+   status = "disabled";
+   clocks = <&clkc 33>,
+     <&clkc 60>,
+     <&clkc 2>;
+   clock-names = "core", "clkin0", "clkin1";
+   resets = <&reset 44>;
+  };
+
+  sd_emmc_b: sd@ffe05000 {
+   compatible = "amlogic,meson-axg-mmc";
+   reg = <0x0 0xffe05000 0x0 0x800>;
+   interrupts = <0 190 4>;
+   status = "disabled";
+   clocks = <&clkc 34>,
+     <&clkc 61>,
+     <&clkc 2>;
+   clock-names = "core", "clkin0", "clkin1";
+   resets = <&reset 45>;
+  };
+
+  sd_emmc_c: mmc@ffe07000 {
+   compatible = "amlogic,meson-axg-mmc";
+   reg = <0x0 0xffe07000 0x0 0x800>;
+   interrupts = <0 191 4>;
+   status = "disabled";
+   clocks = <&clkc 35>,
+     <&clkc 62>,
+     <&clkc 2>;
+   clock-names = "core", "clkin0", "clkin1";
+   resets = <&reset 46>;
+  };
+
+  usb: usb@ffe09000 {
+   status = "disabled";
+   compatible = "amlogic,meson-g12a-usb-ctrl";
+   reg = <0x0 0xffe09000 0x0 0xa0>;
+   interrupts = <0 16 4>;
+   #address-cells = <2>;
+   #size-cells = <2>;
+   ranges;
+
+   clocks = <&clkc 47>;
+   resets = <&reset 34>;
+
+   dr_mode = "otg";
+
+   phys = <&usb2_phy0>, <&usb2_phy1>,
+          <&usb3_pcie_phy 4>;
+   phy-names = "usb2-phy0", "usb2-phy1", "usb3-phy0";
+
+   dwc2: usb@ff400000 {
+    compatible = "amlogic,meson-g12a-usb", "snps,dwc2";
+    reg = <0x0 0xff400000 0x0 0x40000>;
+    interrupts = <0 31 4>;
+    clocks = <&clkc 55>;
+    clock-names = "otg";
+    phys = <&usb2_phy1>;
+    phy-names = "usb2-phy";
+    dr_mode = "peripheral";
+    g-rx-fifo-size = <192>;
+    g-np-tx-fifo-size = <128>;
+    g-tx-fifo-size = <128 128 16 16 16>;
+   };
+
+   dwc3: usb@ff500000 {
+    compatible = "snps,dwc3";
+    reg = <0x0 0xff500000 0x0 0x100000>;
+    interrupts = <0 30 4>;
+    dr_mode = "host";
+    snps,dis_u2_susphy_quirk;
+    snps,quirk-frame-length-adjustment = <0x20>;
+    snps,parkmode-disable-ss-quirk;
+   };
+  };
+
+  mali: gpu@ffe40000 {
+   compatible = "amlogic,meson-g12a-mali", "arm,mali-bifrost";
+   reg = <0x0 0xffe40000 0x0 0x40000>;
+   interrupt-parent = <&gic>;
+   interrupts = <0 162 4>,
+         <0 161 4>,
+         <0 160 4>;
+   interrupt-names = "job", "mmu", "gpu";
+   clocks = <&clkc 175>;
+   resets = <&reset 20>, <&reset 78>;
+   operating-points-v2 = <&gpu_opp_table>;
+   #cooling-cells = <2>;
+  };
+ };
+
+ thermal-zones {
+  cpu_thermal: cpu-thermal {
+   polling-delay = <1000>;
+   polling-delay-passive = <100>;
+   thermal-sensors = <&cpu_temp>;
+
+   trips {
+    cpu_passive: cpu-passive {
+     temperature = <85000>;
+     hysteresis = <2000>;
+     type = "passive";
+    };
+
+    cpu_hot: cpu-hot {
+     temperature = <95000>;
+     hysteresis = <2000>;
+     type = "hot";
+    };
+
+    cpu_critical: cpu-critical {
+     temperature = <110000>;
+     hysteresis = <2000>;
+     type = "critical";
+    };
+   };
+  };
+
+  ddr_thermal: ddr-thermal {
+   polling-delay = <1000>;
+   polling-delay-passive = <100>;
+   thermal-sensors = <&ddr_temp>;
+
+   trips {
+    ddr_passive: ddr-passive {
+     temperature = <85000>;
+     hysteresis = <2000>;
+     type = "passive";
+    };
+
+    ddr_critical: ddr-critical {
+     temperature = <110000>;
+     hysteresis = <2000>;
+     type = "critical";
+    };
+   };
+
+   cooling-maps {
+    map {
+     trip = <&ddr_passive>;
+     cooling-device = <&mali (~0) (~0)>;
+    };
+   };
+  };
+ };
+
+ timer {
+  compatible = "arm,armv8-timer";
+  interrupts = <1 13
+   (((0xff) << 8) | 8)>,
+        <1 14
+   (((0xff) << 8) | 8)>,
+        <1 11
+   (((0xff) << 8) | 8)>,
+        <1 10
+   (((0xff) << 8) | 8)>;
+  arm,no-tick-in-suspend;
+ };
+
+ xtal: xtal-clk {
+  compatible = "fixed-clock";
+  clock-frequency = <24000000>;
+  clock-output-names = "xtal";
+  #clock-cells = <0>;
+ };
+
+ npu: npu@ff100000 {
+  status = "disabled";
+  compatible = "vivante,gc";
+  reg = <0x0 0xff100000 0x0 0x20000>;
+  interrupts = <0 147 4>;
+  clocks = <&clkc 267>,
+    <&clkc 264>;
+  clock-names = "core", "bus";
+  resets = <&reset 76>;
+ };
+};
+# 8 "arch/arm64/boot/dts/amlogic/meson-g12.dtsi" 2
+# 1 "include/dt-bindings/clock/axg-audio-clkc.h" 1
+# 9 "arch/arm64/boot/dts/amlogic/meson-g12.dtsi" 2
+# 1 "include/dt-bindings/power/meson-g12a-power.h" 1
+# 10 "arch/arm64/boot/dts/amlogic/meson-g12.dtsi" 2
+# 1 "include/dt-bindings/reset/amlogic,meson-axg-audio-arb.h" 1
+# 11 "arch/arm64/boot/dts/amlogic/meson-g12.dtsi" 2
+# 1 "include/dt-bindings/reset/amlogic,meson-g12a-audio-reset.h" 1
+# 12 "arch/arm64/boot/dts/amlogic/meson-g12.dtsi" 2
+
+/ {
+ tdmif_a: audio-controller-0 {
+  compatible = "amlogic,axg-tdm-iface";
+  #sound-dai-cells = <0>;
+  sound-name-prefix = "TDM_A";
+  clocks = <&clkc_audio 49>,
+    <&clkc_audio 79>,
+    <&clkc_audio 86>;
+  clock-names = "mclk", "sclk", "lrclk";
+  status = "disabled";
+ };
+
+ tdmif_b: audio-controller-1 {
+  compatible = "amlogic,axg-tdm-iface";
+  #sound-dai-cells = <0>;
+  sound-name-prefix = "TDM_B";
+  clocks = <&clkc_audio 50>,
+    <&clkc_audio 80>,
+    <&clkc_audio 87>;
+  clock-names = "mclk", "sclk", "lrclk";
+  status = "disabled";
+ };
+
+ tdmif_c: audio-controller-2 {
+  compatible = "amlogic,axg-tdm-iface";
+  #sound-dai-cells = <0>;
+  sound-name-prefix = "TDM_C";
+  clocks = <&clkc_audio 51>,
+    <&clkc_audio 81>,
+    <&clkc_audio 88>;
+  clock-names = "mclk", "sclk", "lrclk";
+  status = "disabled";
+ };
+};
+
+&apb {
+ pdm: audio-controller@40000 {
+  compatible = "amlogic,g12a-pdm",
+        "amlogic,axg-pdm";
+  reg = <0x0 0x40000 0x0 0x34>;
+  #sound-dai-cells = <0>;
+  sound-name-prefix = "PDM";
+  clocks = <&clkc_audio 30>,
+    <&clkc_audio 57>,
+    <&clkc_audio 58>;
+  clock-names = "pclk", "dclk", "sysclk";
+  resets = <&clkc_audio 0>;
+  status = "disabled";
+ };
+
+ audio: bus@42000 {
+  compatible = "simple-bus";
+  reg = <0x0 0x42000 0x0 0x2000>;
+  #address-cells = <2>;
+  #size-cells = <2>;
+  ranges = <0x0 0x0 0x0 0x42000 0x0 0x2000>;
+
+  clkc_audio: clock-controller@0 {
+   status = "disabled";
+   compatible = "amlogic,g12a-audio-clkc";
+   reg = <0x0 0x0 0x0 0xb4>;
+   #clock-cells = <1>;
+   #reset-cells = <1>;
+
+   clocks = <&clkc 37>,
+     <&clkc 11>,
+     <&clkc 12>,
+     <&clkc 13>,
+     <&clkc 14>,
+     <&clkc 74>,
+     <&clkc 3>,
+     <&clkc 4>,
+     <&clkc 7>;
+   clock-names = "pclk",
+          "mst_in0",
+          "mst_in1",
+          "mst_in2",
+          "mst_in3",
+          "mst_in4",
+          "mst_in5",
+          "mst_in6",
+          "mst_in7";
+
+   resets = <&reset 65>;
+  };
+
+  toddr_a: audio-controller@100 {
+   compatible = "amlogic,g12a-toddr",
+         "amlogic,axg-toddr";
+   reg = <0x0 0x100 0x0 0x2c>;
+   #sound-dai-cells = <0>;
+   sound-name-prefix = "TODDR_A";
+   interrupts = <0 148 1>;
+   clocks = <&clkc_audio 41>;
+   resets = <&arb 0>,
+     <&clkc_audio 6>;
+   reset-names = "arb", "rst";
+   amlogic,fifo-depth = <512>;
+   status = "disabled";
+  };
+
+  toddr_b: audio-controller@140 {
+   compatible = "amlogic,g12a-toddr",
+         "amlogic,axg-toddr";
+   reg = <0x0 0x140 0x0 0x2c>;
+   #sound-dai-cells = <0>;
+   sound-name-prefix = "TODDR_B";
+   interrupts = <0 149 1>;
+   clocks = <&clkc_audio 42>;
+   resets = <&arb 1>,
+     <&clkc_audio 7>;
+   reset-names = "arb", "rst";
+   amlogic,fifo-depth = <256>;
+   status = "disabled";
+  };
+
+  toddr_c: audio-controller@180 {
+   compatible = "amlogic,g12a-toddr",
+         "amlogic,axg-toddr";
+   reg = <0x0 0x180 0x0 0x2c>;
+   #sound-dai-cells = <0>;
+   sound-name-prefix = "TODDR_C";
+   interrupts = <0 150 1>;
+   clocks = <&clkc_audio 43>;
+   resets = <&arb 2>,
+     <&clkc_audio 8>;
+   reset-names = "arb", "rst";
+   amlogic,fifo-depth = <256>;
+   status = "disabled";
+  };
+
+  frddr_a: audio-controller@1c0 {
+   compatible = "amlogic,g12a-frddr",
+         "amlogic,axg-frddr";
+   reg = <0x0 0x1c0 0x0 0x2c>;
+   #sound-dai-cells = <0>;
+   sound-name-prefix = "FRDDR_A";
+   interrupts = <0 152 1>;
+   clocks = <&clkc_audio 38>;
+   resets = <&arb 3>,
+     <&clkc_audio 9>;
+   reset-names = "arb", "rst";
+   amlogic,fifo-depth = <512>;
+   status = "disabled";
+  };
+
+  frddr_b: audio-controller@200 {
+   compatible = "amlogic,g12a-frddr",
+         "amlogic,axg-frddr";
+   reg = <0x0 0x200 0x0 0x2c>;
+   #sound-dai-cells = <0>;
+   sound-name-prefix = "FRDDR_B";
+   interrupts = <0 153 1>;
+   clocks = <&clkc_audio 39>;
+   resets = <&arb 4>,
+     <&clkc_audio 10>;
+   reset-names = "arb", "rst";
+   amlogic,fifo-depth = <256>;
+   status = "disabled";
+  };
+
+  frddr_c: audio-controller@240 {
+   compatible = "amlogic,g12a-frddr",
+         "amlogic,axg-frddr";
+   reg = <0x0 0x240 0x0 0x2c>;
+   #sound-dai-cells = <0>;
+   sound-name-prefix = "FRDDR_C";
+   interrupts = <0 154 1>;
+   clocks = <&clkc_audio 40>;
+   resets = <&arb 5>,
+     <&clkc_audio 11>;
+   reset-names = "arb", "rst";
+   amlogic,fifo-depth = <256>;
+   status = "disabled";
+  };
+
+  arb: reset-controller@280 {
+   status = "disabled";
+   compatible = "amlogic,meson-axg-audio-arb";
+   reg = <0x0 0x280 0x0 0x4>;
+   #reset-cells = <1>;
+   clocks = <&clkc_audio 29>;
+  };
+
+  tdmin_a: audio-controller@300 {
+   compatible = "amlogic,g12a-tdmin",
+         "amlogic,axg-tdmin";
+   reg = <0x0 0x300 0x0 0x40>;
+   sound-name-prefix = "TDMIN_A";
+   resets = <&clkc_audio 1>;
+   clocks = <&clkc_audio 31>,
+     <&clkc_audio 123>,
+     <&clkc_audio 116>,
+     <&clkc_audio 130>,
+     <&clkc_audio 130>;
+   clock-names = "pclk", "sclk", "sclk_sel",
+          "lrclk", "lrclk_sel";
+   status = "disabled";
+  };
+
+  tdmin_b: audio-controller@340 {
+   compatible = "amlogic,g12a-tdmin",
+         "amlogic,axg-tdmin";
+   reg = <0x0 0x340 0x0 0x40>;
+   sound-name-prefix = "TDMIN_B";
+   resets = <&clkc_audio 2>;
+   clocks = <&clkc_audio 32>,
+     <&clkc_audio 124>,
+     <&clkc_audio 117>,
+     <&clkc_audio 131>,
+     <&clkc_audio 131>;
+   clock-names = "pclk", "sclk", "sclk_sel",
+          "lrclk", "lrclk_sel";
+   status = "disabled";
+  };
+
+  tdmin_c: audio-controller@380 {
+   compatible = "amlogic,g12a-tdmin",
+         "amlogic,axg-tdmin";
+   reg = <0x0 0x380 0x0 0x40>;
+   sound-name-prefix = "TDMIN_C";
+   resets = <&clkc_audio 3>;
+   clocks = <&clkc_audio 33>,
+     <&clkc_audio 125>,
+     <&clkc_audio 118>,
+     <&clkc_audio 132>,
+     <&clkc_audio 132>;
+   clock-names = "pclk", "sclk", "sclk_sel",
+          "lrclk", "lrclk_sel";
+   status = "disabled";
+  };
+
+  tdmin_lb: audio-controller@3c0 {
+   compatible = "amlogic,g12a-tdmin",
+         "amlogic,axg-tdmin";
+   reg = <0x0 0x3c0 0x0 0x40>;
+   sound-name-prefix = "TDMIN_LB";
+   resets = <&clkc_audio 4>;
+   clocks = <&clkc_audio 34>,
+     <&clkc_audio 126>,
+     <&clkc_audio 119>,
+     <&clkc_audio 133>,
+     <&clkc_audio 133>;
+   clock-names = "pclk", "sclk", "sclk_sel",
+          "lrclk", "lrclk_sel";
+   status = "disabled";
+  };
+
+  spdifin: audio-controller@400 {
+   compatible = "amlogic,g12a-spdifin",
+         "amlogic,axg-spdifin";
+   reg = <0x0 0x400 0x0 0x30>;
+   #sound-dai-cells = <0>;
+   sound-name-prefix = "SPDIFIN";
+   interrupts = <0 151 1>;
+   clocks = <&clkc_audio 45>,
+     <&clkc_audio 56>;
+   clock-names = "pclk", "refclk";
+   resets = <&clkc_audio 17>;
+   status = "disabled";
+  };
+
+  spdifout: audio-controller@480 {
+   compatible = "amlogic,g12a-spdifout",
+         "amlogic,axg-spdifout";
+   reg = <0x0 0x480 0x0 0x50>;
+   #sound-dai-cells = <0>;
+   sound-name-prefix = "SPDIFOUT";
+   clocks = <&clkc_audio 46>,
+     <&clkc_audio 55>;
+   clock-names = "pclk", "mclk";
+   resets = <&clkc_audio 15>;
+   status = "disabled";
+  };
+
+  tdmout_a: audio-controller@500 {
+   compatible = "amlogic,g12a-tdmout";
+   reg = <0x0 0x500 0x0 0x40>;
+   sound-name-prefix = "TDMOUT_A";
+   resets = <&clkc_audio 12>;
+   clocks = <&clkc_audio 35>,
+     <&clkc_audio 127>,
+     <&clkc_audio 120>,
+     <&clkc_audio 134>,
+     <&clkc_audio 134>;
+   clock-names = "pclk", "sclk", "sclk_sel",
+          "lrclk", "lrclk_sel";
+   status = "disabled";
+  };
+
+  tdmout_b: audio-controller@540 {
+   compatible = "amlogic,g12a-tdmout";
+   reg = <0x0 0x540 0x0 0x40>;
+   sound-name-prefix = "TDMOUT_B";
+   resets = <&clkc_audio 13>;
+   clocks = <&clkc_audio 36>,
+     <&clkc_audio 128>,
+     <&clkc_audio 121>,
+     <&clkc_audio 135>,
+     <&clkc_audio 135>;
+   clock-names = "pclk", "sclk", "sclk_sel",
+          "lrclk", "lrclk_sel";
+   status = "disabled";
+  };
+
+  tdmout_c: audio-controller@580 {
+   compatible = "amlogic,g12a-tdmout";
+   reg = <0x0 0x580 0x0 0x40>;
+   sound-name-prefix = "TDMOUT_C";
+   resets = <&clkc_audio 14>;
+   clocks = <&clkc_audio 37>,
+     <&clkc_audio 129>,
+     <&clkc_audio 122>,
+     <&clkc_audio 136>,
+     <&clkc_audio 136>;
+   clock-names = "pclk", "sclk", "sclk_sel",
+          "lrclk", "lrclk_sel";
+   status = "disabled";
+  };
+
+  spdifout_b: audio-controller@680 {
+   compatible = "amlogic,g12a-spdifout",
+         "amlogic,axg-spdifout";
+   reg = <0x0 0x680 0x0 0x50>;
+   #sound-dai-cells = <0>;
+   sound-name-prefix = "SPDIFOUT_B";
+   clocks = <&clkc_audio 151>,
+     <&clkc_audio 152>;
+   clock-names = "pclk", "mclk";
+   resets = <&clkc_audio 16>;
+   status = "disabled";
+  };
+
+  toacodec: audio-controller@740 {
+   compatible = "amlogic,g12a-toacodec";
+   reg = <0x0 0x740 0x0 0x4>;
+   #sound-dai-cells = <1>;
+   sound-name-prefix = "TOACODEC";
+   resets = <&clkc_audio 23>;
+   status = "disabled";
+  };
+
+  tohdmitx: audio-controller@744 {
+   compatible = "amlogic,g12a-tohdmitx";
+   reg = <0x0 0x744 0x0 0x4>;
+   #sound-dai-cells = <1>;
+   sound-name-prefix = "TOHDMITX";
+   resets = <&clkc_audio 24>;
+   status = "disabled";
+  };
+ };
+};
+
+&ethmac {
+ power-domains = <&pwrc 1>;
+};
+
+&vpu {
+ power-domains = <&pwrc 0>;
+};
+
+&sd_emmc_a {
+ amlogic,dram-access-quirk;
+};
+
+&simplefb_cvbs {
+ power-domains = <&pwrc 0>;
+};
+
+&simplefb_hdmi {
+ power-domains = <&pwrc 0>;
+};
+# 8 "arch/arm64/boot/dts/amlogic/meson-g12b.dtsi" 2
+
+/ {
+ compatible = "amlogic,g12b";
+
+ cpus {
+  #address-cells = <0x2>;
+  #size-cells = <0x0>;
+
+  cpu-map {
+   cluster0 {
+    core0 {
+     cpu = <&cpu0>;
+    };
+
+    core1 {
+     cpu = <&cpu1>;
+    };
+   };
+
+   cluster1 {
+    core0 {
+     cpu = <&cpu100>;
+    };
+
+    core1 {
+     cpu = <&cpu101>;
+    };
+
+    core2 {
+     cpu = <&cpu102>;
+    };
+
+    core3 {
+     cpu = <&cpu103>;
+    };
+   };
+  };
+
+  cpu0: cpu@0 {
+   device_type = "cpu";
+   compatible = "arm,cortex-a53";
+   reg = <0x0 0x0>;
+   enable-method = "psci";
+   capacity-dmips-mhz = <592>;
+   next-level-cache = <&l2>;
+   #cooling-cells = <2>;
+  };
+
+  cpu1: cpu@1 {
+   device_type = "cpu";
+   compatible = "arm,cortex-a53";
+   reg = <0x0 0x1>;
+   enable-method = "psci";
+   capacity-dmips-mhz = <592>;
+   next-level-cache = <&l2>;
+   #cooling-cells = <2>;
+  };
+
+  cpu100: cpu@100 {
+   device_type = "cpu";
+   compatible = "arm,cortex-a73";
+   reg = <0x0 0x100>;
+   enable-method = "psci";
+   capacity-dmips-mhz = <1024>;
+   next-level-cache = <&l2>;
+   #cooling-cells = <2>;
+  };
+
+  cpu101: cpu@101 {
+   device_type = "cpu";
+   compatible = "arm,cortex-a73";
+   reg = <0x0 0x101>;
+   enable-method = "psci";
+   capacity-dmips-mhz = <1024>;
+   next-level-cache = <&l2>;
+   #cooling-cells = <2>;
+  };
+
+  cpu102: cpu@102 {
+   device_type = "cpu";
+   compatible = "arm,cortex-a73";
+   reg = <0x0 0x102>;
+   enable-method = "psci";
+   capacity-dmips-mhz = <1024>;
+   next-level-cache = <&l2>;
+   #cooling-cells = <2>;
+  };
+
+  cpu103: cpu@103 {
+   device_type = "cpu";
+   compatible = "arm,cortex-a73";
+   reg = <0x0 0x103>;
+   enable-method = "psci";
+   capacity-dmips-mhz = <1024>;
+   next-level-cache = <&l2>;
+   #cooling-cells = <2>;
+  };
+
+  l2: l2-cache0 {
+   compatible = "cache";
+  };
+ };
+};
+
+&clkc {
+ compatible = "amlogic,g12b-clkc";
+};
+
+&cpu_thermal {
+ cooling-maps {
+  map0 {
+   trip = <&cpu_passive>;
+   cooling-device = <&cpu0 (~0) (~0)>,
+      <&cpu1 (~0) (~0)>,
+      <&cpu100 (~0) (~0)>,
+      <&cpu101 (~0) (~0)>,
+      <&cpu102 (~0) (~0)>,
+      <&cpu103 (~0) (~0)>;
+  };
+  map1 {
+   trip = <&cpu_hot>;
+   cooling-device = <&cpu0 (~0) (~0)>,
+      <&cpu1 (~0) (~0)>,
+      <&cpu100 (~0) (~0)>,
+      <&cpu101 (~0) (~0)>,
+      <&cpu102 (~0) (~0)>,
+      <&cpu103 (~0) (~0)>;
+  };
+ };
+};
+
+&mali {
+ dma-coherent;
+};
+
+&npu {
+ power-domains = <&pwrc 2>;
+};
+# 8 "arch/arm64/boot/dts/amlogic/meson-g12b-a311d.dtsi" 2
+
+/ {
+ cpu_opp_table_0: opp-table-0 {
+  compatible = "operating-points-v2";
+  opp-shared;
+
+  opp-1000000000 {
+   opp-hz = /bits/ 64 <1000000000>;
+   opp-microvolt = <761000>;
+  };
+
+  opp-1200000000 {
+   opp-hz = /bits/ 64 <1200000000>;
+   opp-microvolt = <781000>;
+  };
+
+  opp-1398000000 {
+   opp-hz = /bits/ 64 <1398000000>;
+   opp-microvolt = <811000>;
+  };
+
+  opp-1512000000 {
+   opp-hz = /bits/ 64 <1512000000>;
+   opp-microvolt = <861000>;
+  };
+
+  opp-1608000000 {
+   opp-hz = /bits/ 64 <1608000000>;
+   opp-microvolt = <901000>;
+  };
+
+  opp-1704000000 {
+   opp-hz = /bits/ 64 <1704000000>;
+   opp-microvolt = <951000>;
+  };
+
+  opp-1800000000 {
+   opp-hz = /bits/ 64 <1800000000>;
+   opp-microvolt = <1001000>;
+  };
+ };
+
+ cpub_opp_table_1: opp-table-1 {
+  compatible = "operating-points-v2";
+  opp-shared;
+
+  opp-1000000000 {
+   opp-hz = /bits/ 64 <1000000000>;
+   opp-microvolt = <731000>;
+  };
+
+  opp-1200000000 {
+   opp-hz = /bits/ 64 <1200000000>;
+   opp-microvolt = <751000>;
+  };
+
+  opp-1398000000 {
+   opp-hz = /bits/ 64 <1398000000>;
+   opp-microvolt = <771000>;
+  };
+
+  opp-1512000000 {
+   opp-hz = /bits/ 64 <1512000000>;
+   opp-microvolt = <771000>;
+  };
+
+  opp-1608000000 {
+   opp-hz = /bits/ 64 <1608000000>;
+   opp-microvolt = <781000>;
+  };
+
+  opp-1704000000 {
+   opp-hz = /bits/ 64 <1704000000>;
+   opp-microvolt = <791000>;
+  };
+
+  opp-1800000000 {
+   opp-hz = /bits/ 64 <1800000000>;
+   opp-microvolt = <831000>;
+  };
+
+                opp-1908000000 {
+                        opp-hz = /bits/ 64 <1908000000>;
+                        opp-microvolt = <861000>;
+                };
+
+                opp-2016000000 {
+                        opp-hz = /bits/ 64 <2016000000>;
+                        opp-microvolt = <911000>;
+                };
+
+                opp-2108000000 {
+                        opp-hz = /bits/ 64 <2108000000>;
+                        opp-microvolt = <951000>;
+                };
+
+                opp-2208000000 {
+                        opp-hz = /bits/ 64 <2208000000>;
+                        opp-microvolt = <1011000>;
+                };
+ };
+};
+# 10 "arch/arm64/boot/dts/amlogic/meson-g12b-odroid-n2l.dts" 2
+# 1 "arch/arm64/boot/dts/amlogic/meson-g12b-odroid.dtsi" 1
+
+
+
+
+
+
+# 1 "include/dt-bindings/input/input.h" 1
+# 13 "include/dt-bindings/input/input.h"
+# 1 "include/dt-bindings/input/linux-event-codes.h" 1
+# 14 "include/dt-bindings/input/input.h" 2
+# 8 "arch/arm64/boot/dts/amlogic/meson-g12b-odroid.dtsi" 2
+# 1 "include/dt-bindings/gpio/meson-g12a-gpio.h" 1
+# 9 "arch/arm64/boot/dts/amlogic/meson-g12b-odroid.dtsi" 2
+# 1 "include/dt-bindings/sound/meson-g12a-toacodec.h" 1
+# 10 "arch/arm64/boot/dts/amlogic/meson-g12b-odroid.dtsi" 2
+# 1 "include/dt-bindings/sound/meson-g12a-tohdmitx.h" 1
+# 11 "arch/arm64/boot/dts/amlogic/meson-g12b-odroid.dtsi" 2
+
+/ {
+ aliases {
+  serial0 = &uart_AO;
+  ethernet0 = &ethmac;
+  rtc1 = &vrtc;
+ };
+
+ chosen {
+  stdout-path = "serial0:115200n8";
+ };
+
+ memory@0 {
+  device_type = "memory";
+  reg = <0x0 0x0 0x0 0x40000000>;
+ };
+
+ emmc_pwrseq: emmc-pwrseq {
+  compatible = "mmc-pwrseq-emmc";
+  reset-gpios = <&gpio 37 1>;
+ };
+
+ fan: gpio-fan {
+  compatible = "gpio-fan";
+  gpios = <&gpio_ao 10 0>;
+
+  gpio-fan,speed-map = <0 0>, <1 1>;
+  #cooling-cells = <2>;
+ };
+
+ leds {
+  compatible = "gpio-leds";
+
+  led-blue {
+   label = "n2:blue";
+   gpios = <&gpio_ao 11 0>;
+   linux,default-trigger = "heartbeat";
+  };
+ };
+
+ tflash_vdd: regulator-tflash_vdd {
+  compatible = "regulator-fixed";
+
+  regulator-name = "TFLASH_VDD";
+  regulator-min-microvolt = <3300000>;
+  regulator-max-microvolt = <3300000>;
+
+  gpio = <&gpio_ao 8 0>;
+  enable-active-high;
+  regulator-always-on;
+ };
+
+ tf_io: gpio-regulator-tf_io {
+  compatible = "regulator-gpio";
+
+  regulator-name = "TF_IO";
+  regulator-min-microvolt = <1800000>;
+  regulator-max-microvolt = <3300000>;
+
+  gpios = <&gpio_ao 9 0>;
+  gpios-states = <0>;
+
+  states = <3300000 0>,
+    <1800000 1>;
+ };
+
+ flash_1v8: regulator-flash_1v8 {
+  compatible = "regulator-fixed";
+  regulator-name = "FLASH_1V8";
+  regulator-min-microvolt = <1800000>;
+  regulator-max-microvolt = <1800000>;
+  vin-supply = <&vcc_3v3>;
+  regulator-always-on;
+ };
+
+ main_12v: regulator-main_12v {
+  compatible = "regulator-fixed";
+  regulator-name = "12V";
+  regulator-min-microvolt = <12000000>;
+  regulator-max-microvolt = <12000000>;
+  regulator-always-on;
+ };
+
+ usb_pwr_en: regulator-usb_pwr_en {
+  compatible = "regulator-fixed";
+  regulator-name = "USB_PWR_EN";
+  regulator-min-microvolt = <5000000>;
+  regulator-max-microvolt = <5000000>;
+  vin-supply = <&vcc_5v>;
+
+
+  gpio = <&gpio 22 0>;
+  enable-active-high;
+ };
+
+ vcc_5v: regulator-vcc_5v {
+  compatible = "regulator-fixed";
+  regulator-name = "5V";
+  regulator-min-microvolt = <5000000>;
+  regulator-max-microvolt = <5000000>;
+  regulator-always-on;
+  vin-supply = <&main_12v>;
+  gpio = <&gpio 24 (2 | 4)>;
+  enable-active-high;
+ };
+
+ vcc_1v8: regulator-vcc_1v8 {
+  compatible = "regulator-fixed";
+  regulator-name = "VCC_1V8";
+  regulator-min-microvolt = <1800000>;
+  regulator-max-microvolt = <1800000>;
+  vin-supply = <&vcc_3v3>;
+  regulator-always-on;
+ };
+
+ vcc_3v3: regulator-vcc_3v3 {
+  compatible = "regulator-fixed";
+  regulator-name = "VCC_3V3";
+  regulator-min-microvolt = <3300000>;
+  regulator-max-microvolt = <3300000>;
+  vin-supply = <&vddao_3v3>;
+  regulator-always-on;
+
+ };
+
+ vddcpu_a: regulator-vddcpu-a {
+
+
+
+  compatible = "pwm-regulator";
+
+  regulator-name = "VDDCPU_A";
+  regulator-min-microvolt = <721000>;
+  regulator-max-microvolt = <1022000>;
+
+  pwm-supply = <&main_12v>;
+
+  pwms = <&pwm_ab 0 1250 0>;
+  pwm-dutycycle-range = <100 0>;
+
+  regulator-boot-on;
+  regulator-always-on;
+ };
+
+ vddcpu_b: regulator-vddcpu-b {
+
+
+
+  compatible = "pwm-regulator";
+
+  regulator-name = "VDDCPU_B";
+  regulator-min-microvolt = <721000>;
+  regulator-max-microvolt = <1022000>;
+
+  pwm-supply = <&main_12v>;
+
+  pwms = <&pwm_AO_cd 1 1250 0>;
+  pwm-dutycycle-range = <100 0>;
+
+  regulator-boot-on;
+  regulator-always-on;
+ };
+
+ vddao_1v8: regulator-vddao_1v8 {
+  compatible = "regulator-fixed";
+  regulator-name = "VDDAO_1V8";
+  regulator-min-microvolt = <1800000>;
+  regulator-max-microvolt = <1800000>;
+  vin-supply = <&vddao_3v3>;
+  regulator-always-on;
+ };
+
+ vddao_3v3: regulator-vddao_3v3 {
+  compatible = "regulator-fixed";
+  regulator-name = "VDDAO_3V3";
+  regulator-min-microvolt = <3300000>;
+  regulator-max-microvolt = <3300000>;
+  vin-supply = <&main_12v>;
+  regulator-always-on;
+ };
+
+ hdmi-connector {
+  compatible = "hdmi-connector";
+  type = "a";
+
+  port {
+   hdmi_connector_in: endpoint {
+    remote-endpoint = <&hdmi_tx_tmds_out>;
+   };
+  };
+ };
+};
+
+&arb {
+ status = "okay";
+};
+
+&cec_AO {
+ pinctrl-0 = <&cec_ao_a_h_pins>;
+ pinctrl-names = "default";
+ status = "disabled";
+ hdmi-phandle = <&hdmi_tx>;
+};
+
+&cecb_AO {
+ pinctrl-0 = <&cec_ao_b_h_pins>;
+ pinctrl-names = "default";
+ status = "okay";
+ hdmi-phandle = <&hdmi_tx>;
+};
+
+&clkc_audio {
+ status = "okay";
+};
+
+&cpu0 {
+ cpu-supply = <&vddcpu_b>;
+ operating-points-v2 = <&cpu_opp_table_0>;
+ clocks = <&clkc 187>;
+ clock-latency = <50000>;
+};
+
+&cpu1 {
+ cpu-supply = <&vddcpu_b>;
+ operating-points-v2 = <&cpu_opp_table_0>;
+ clocks = <&clkc 187>;
+ clock-latency = <50000>;
+};
+
+&cpu100 {
+ cpu-supply = <&vddcpu_a>;
+ operating-points-v2 = <&cpub_opp_table_1>;
+ clocks = <&clkc 224>;
+ clock-latency = <50000>;
+};
+
+&cpu101 {
+ cpu-supply = <&vddcpu_a>;
+ operating-points-v2 = <&cpub_opp_table_1>;
+ clocks = <&clkc 224>;
+ clock-latency = <50000>;
+};
+
+&cpu102 {
+ cpu-supply = <&vddcpu_a>;
+ operating-points-v2 = <&cpub_opp_table_1>;
+ clocks = <&clkc 224>;
+ clock-latency = <50000>;
+};
+
+&cpu103 {
+ cpu-supply = <&vddcpu_a>;
+ operating-points-v2 = <&cpub_opp_table_1>;
+ clocks = <&clkc 224>;
+ clock-latency = <50000>;
+};
+
+&cpu_thermal {
+ trips {
+  cpu_active: cpu-active {
+   temperature = <60000>;
+   hysteresis = <2000>;
+   type = "active";
+  };
+ };
+
+ cooling-maps {
+  map {
+   trip = <&cpu_active>;
+   cooling-device = <&fan (~0) (~0)>;
+  };
+ };
+};
+
+&ddr_thermal {
+ trips {
+  ddr_active: ddr-active {
+   temperature = <60000>;
+   hysteresis = <2000>;
+   type = "active";
+  };
+ };
+
+ cooling-maps {
+  map {
+   trip = <&ddr_active>;
+   cooling-device = <&fan (~0) (~0)>;
+  };
+ };
+};
+
+&frddr_a {
+ status = "okay";
+};
+
+&frddr_b {
+ status = "okay";
+};
+
+&frddr_c {
+ status = "okay";
+};
+
+&hdmi_tx {
+ status = "okay";
+ pinctrl-0 = <&hdmitx_hpd_pins>, <&hdmitx_ddc_pins>;
+ pinctrl-names = "default";
+ hdmi-supply = <&vcc_5v>;
+};
+
+&hdmi_tx_tmds_port {
+ hdmi_tx_tmds_out: endpoint {
+  remote-endpoint = <&hdmi_connector_in>;
+ };
+};
+
+&pwm_ab {
+ pinctrl-0 = <&pwm_a_e_pins>;
+ pinctrl-names = "default";
+ clocks = <&xtal>;
+ clock-names = "clkin0";
+ status = "okay";
+};
+
+&pwm_AO_cd {
+ pinctrl-0 = <&pwm_ao_d_e_pins>;
+ pinctrl-names = "default";
+ clocks = <&xtal>;
+ clock-names = "clkin1";
+ status = "okay";
+};
+
+&saradc {
+ status = "okay";
+ vref-supply = <&vddao_1v8>;
+};
+
+
+&sd_emmc_b {
+ status = "okay";
+ pinctrl-0 = <&sdcard_c_pins>;
+ pinctrl-1 = <&sdcard_clk_gate_c_pins>;
+ pinctrl-names = "default", "clk-gate";
+
+ bus-width = <4>;
+ cap-sd-highspeed;
+ max-frequency = <50000000>;
+ disable-wp;
+
+ cd-gpios = <&gpio 47 1>;
+ vmmc-supply = <&tflash_vdd>;
+ vqmmc-supply = <&tf_io>;
+
+};
+
+
+&sd_emmc_c {
+ status = "okay";
+ pinctrl-0 = <&emmc_ctrl_pins>, <&emmc_data_8b_pins>, <&emmc_ds_pins>;
+ pinctrl-1 = <&emmc_clk_gate_pins>;
+ pinctrl-names = "default", "clk-gate";
+
+ bus-width = <8>;
+ cap-mmc-highspeed;
+ mmc-ddr-1_8v;
+ mmc-hs200-1_8v;
+ max-frequency = <200000000>;
+ disable-wp;
+
+ mmc-pwrseq = <&emmc_pwrseq>;
+ vmmc-supply = <&vcc_3v3>;
+ vqmmc-supply = <&flash_1v8>;
+};
+
+&tdmif_b {
+ status = "okay";
+};
+
+&tdmif_c {
+ status = "okay";
+};
+
+&tdmin_a {
+ status = "okay";
+};
+
+&tdmin_b {
+ status = "okay";
+};
+
+&tdmin_c {
+ status = "okay";
+};
+
+&tdmin_lb {
+ status = "okay";
+};
+
+&tdmout_b {
+ status = "okay";
+};
+
+&tdmout_c {
+ status = "okay";
+};
+
+&tohdmitx {
+ status = "okay";
+};
+
+&toddr_a {
+ status = "okay";
+};
+
+&toddr_b {
+ status = "okay";
+};
+
+&toddr_c {
+ status = "okay";
+};
+
+&uart_AO {
+ status = "okay";
+ pinctrl-0 = <&uart_ao_a_pins>;
+ pinctrl-names = "default";
+};
+
+&usb {
+ status = "okay";
+};
+
+&usb2_phy0 {
+ phy-supply = <&vcc_5v>;
+};
+# 11 "arch/arm64/boot/dts/amlogic/meson-g12b-odroid-n2l.dts" 2
+
+/ {
+ compatible = "hardkernel,odroid-n2l", "amlogic,s922x", "amlogic,g12b";
+ model = "Hardkernel ODROID-N2L";
+
+ sound {
+  compatible = "amlogic,axg-sound-card";
+  model = "ODROID-N2L";
+  audio-aux-devs = <&tdmout_b>, <&tdmin_a>, <&tdmin_b>,
+     <&tdmin_c>, <&tdmin_lb>;
+  audio-routing = "TDMOUT_B IN 0", "FRDDR_A OUT 1",
+    "TDMOUT_B IN 1", "FRDDR_B OUT 1",
+    "TDMOUT_B IN 2", "FRDDR_C OUT 1",
+    "TDM_B Playback", "TDMOUT_B OUT",
+    "TDMIN_A IN 4", "TDM_B Loopback",
+    "TDMIN_B IN 4", "TDM_B Loopback",
+    "TDMIN_C IN 4", "TDM_B Loopback",
+    "TDMIN_LB IN 1", "TDM_B Loopback",
+    "TODDR_A IN 0", "TDMIN_A OUT",
+    "TODDR_B IN 0", "TDMIN_A OUT",
+    "TODDR_C IN 0", "TDMIN_A OUT",
+    "TODDR_A IN 1", "TDMIN_B OUT",
+    "TODDR_B IN 1", "TDMIN_B OUT",
+    "TODDR_C IN 1", "TDMIN_B OUT",
+    "TODDR_A IN 2", "TDMIN_C OUT",
+    "TODDR_B IN 2", "TDMIN_C OUT",
+    "TODDR_C IN 2", "TDMIN_C OUT",
+    "TODDR_A IN 6", "TDMIN_LB OUT",
+    "TODDR_B IN 6", "TDMIN_LB OUT",
+    "TODDR_C IN 6", "TDMIN_LB OUT";
+
+  assigned-clocks = <&clkc 13>,
+      <&clkc 11>,
+      <&clkc 12>;
+  assigned-clock-parents = <0>, <0>, <0>;
+  assigned-clock-rates = <294912000>,
+           <270950400>,
+           <393216000>;
+  status = "okay";
+
+  dai-link-0 {
+   sound-dai = <&frddr_a>;
+  };
+
+  dai-link-1 {
+   sound-dai = <&frddr_b>;
+  };
+
+  dai-link-2 {
+   sound-dai = <&frddr_c>;
+  };
+
+  dai-link-3 {
+   sound-dai = <&toddr_a>;
+  };
+
+  dai-link-4 {
+   sound-dai = <&toddr_b>;
+  };
+
+  dai-link-5 {
+   sound-dai = <&toddr_c>;
+  };
+
+
+  dai-link-6 {
+   sound-dai = <&tdmif_b>;
+   dai-format = "i2s";
+   dai-tdm-slot-tx-mask-0 = <1 1>;
+   dai-tdm-slot-tx-mask-1 = <1 1>;
+   dai-tdm-slot-tx-mask-2 = <1 1>;
+   dai-tdm-slot-tx-mask-3 = <1 1>;
+   mclk-fs = <256>;
+
+   codec {
+    sound-dai = <&tohdmitx 1>;
+   };
+  };
+
+
+  dai-link-7 {
+   sound-dai = <&tohdmitx 3>;
+
+   codec {
+    sound-dai = <&hdmi_tx>;
+   };
+  };
+ };
+};
+
+&eth_phy {
+ status = "disabled";
+};
+
+&vddcpu_a {
+ regulator-min-microvolt = <680000>;
+ regulator-max-microvolt = <1040000>;
+
+ pwms = <&pwm_ab 0 1500 0>;
+};
+
+&vddcpu_b {
+ regulator-min-microvolt = <680000>;
+ regulator-max-microvolt = <1040000>;
+
+ pwms = <&pwm_AO_cd 1 1500 0>;
+};
+
+&usb2_phy0 {
+ phy-supply = <&usb_pwr_en>;
+};
+
+&usb2_phy1 {
+ phy-supply = <&usb_pwr_en>;
+};
-- 
2.30.2

